{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "4bd76c5d",
   "metadata": {},
   "source": [
    "# 搭建卷积神经网络，对手写数字识别图像数据集进行 10 分类。\n",
    "## 算法流程：\n",
    "### （1）读取、加载、处理数据集（可选取部分数据集）； （√）\n",
    "### （2）搭建卷积神经网络； （√）\n",
    "### （3）训练所搭建的神经网络； （√）\n",
    "### （4）利用准确率等指标评估所训练的神经网络；（√）\n",
    "### 加分项 1，保存模型并加载模型，用自己手写的数字图片（手机的便签或微 软画图工具，保证没有阴影）测试分类效果，像素需要降低尽量保持在 28x28x1。 （√）\n",
    "### 加分项 2，使用 pytorch 或只使用 tensorflow 完成两个案例任意一个（tf√）"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9f98af29",
   "metadata": {},
   "source": [
    "（1）读取、加载、处理数据集（可选取部分数据集）；"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "fb5d165c",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Using TensorFlow backend.\n",
      "WARNING:root:Limited tf.compat.v2.summary API due to missing TensorBoard installation.\n"
     ]
    }
   ],
   "source": [
    "from keras.models import Sequential\n",
    "from keras.layers import Conv2D, MaxPool2D\n",
    "from keras.layers import Dense, Flatten\n",
    "import keras\n",
    "from keras.datasets import mnist\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "#加载数据\n",
    "(X_train, y_train), (X_test, y_test) = mnist.load_data()\n",
    "img_x, img_y = X_train.shape[1], X_train.shape[2]\n",
    "\n",
    "#数据预处理\n",
    "X_train = X_train.reshape(X_train.shape[0], img_x, img_y, 1)\n",
    "X_test = X_test.reshape(X_test.shape[0], img_x, img_y, 1)\n",
    "X_train = X_train.astype('float32')\n",
    "X_test = X_test.astype('float32')\n",
    "X_train /= 255\n",
    "X_test /= 255\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e2b9bf07",
   "metadata": {},
   "source": [
    "2）搭建卷积神经网络； （加分项 2，使用 pytorch 或只使用 tensorflow 完成两个案例任意一个）"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "118a845b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "WARNING:tensorflow:From C:\\Users\\29459\\anaconda3\\envs\\python37\\lib\\site-packages\\tensorflow_core\\python\\ops\\resource_variable_ops.py:1630: calling BaseResourceVariable.__init__ (from tensorflow.python.ops.resource_variable_ops) with constraint is deprecated and will be removed in a future version.\n",
      "Instructions for updating:\n",
      "If using Keras pass *_constraint arguments to layers.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "WARNING:tensorflow:From C:\\Users\\29459\\anaconda3\\envs\\python37\\lib\\site-packages\\tensorflow_core\\python\\ops\\resource_variable_ops.py:1630: calling BaseResourceVariable.__init__ (from tensorflow.python.ops.resource_variable_ops) with constraint is deprecated and will be removed in a future version.\n",
      "Instructions for updating:\n",
      "If using Keras pass *_constraint arguments to layers.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "WARNING:tensorflow:From C:\\Users\\29459\\anaconda3\\envs\\python37\\lib\\site-packages\\keras\\backend\\tensorflow_backend.py:4070: The name tf.nn.max_pool is deprecated. Please use tf.nn.max_pool2d instead.\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "WARNING:tensorflow:From C:\\Users\\29459\\anaconda3\\envs\\python37\\lib\\site-packages\\keras\\backend\\tensorflow_backend.py:4070: The name tf.nn.max_pool is deprecated. Please use tf.nn.max_pool2d instead.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "y_train = keras.utils.to_categorical(y_train, num_classes=10)\n",
    "y_test = keras.utils.to_categorical(y_test, num_classes=10)\n",
    "\n",
    "#构建模型\n",
    "model = Sequential()\n",
    "model.add(Conv2D(32, kernel_size=(5,5),padding='same', activation='relu', input_shape=(img_x, img_y, 1)))\n",
    "model.add(MaxPool2D(pool_size=(2,2), strides=(2,2)))\n",
    "model.add(Conv2D(64, kernel_size=(5,5), padding='same',activation='relu'))\n",
    "model.add(MaxPool2D(pool_size=(2,2), strides=(2,2)))\n",
    "model.add(Flatten())\n",
    "model.add(Dense(1000, activation='relu'))\n",
    "model.add(Dense(10, activation='softmax'))\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "7b19ea36",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Model: \"sequential_1\"\n",
      "_________________________________________________________________\n",
      "Layer (type)                 Output Shape              Param #   \n",
      "=================================================================\n",
      "conv2d_1 (Conv2D)            (None, 28, 28, 32)        832       \n",
      "_________________________________________________________________\n",
      "max_pooling2d_1 (MaxPooling2 (None, 14, 14, 32)        0         \n",
      "_________________________________________________________________\n",
      "conv2d_2 (Conv2D)            (None, 14, 14, 64)        51264     \n",
      "_________________________________________________________________\n",
      "max_pooling2d_2 (MaxPooling2 (None, 7, 7, 64)          0         \n",
      "_________________________________________________________________\n",
      "flatten_1 (Flatten)          (None, 3136)              0         \n",
      "_________________________________________________________________\n",
      "dense_1 (Dense)              (None, 1000)              3137000   \n",
      "_________________________________________________________________\n",
      "dense_2 (Dense)              (None, 10)                10010     \n",
      "=================================================================\n",
      "Total params: 3,199,106\n",
      "Trainable params: 3,199,106\n",
      "Non-trainable params: 0\n",
      "_________________________________________________________________\n"
     ]
    }
   ],
   "source": [
    "model.summary()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ccc5f877",
   "metadata": {},
   "source": [
    "（3）训练所搭建的神经网络； （√）"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "ff0fb22e",
   "metadata": {},
   "outputs": [],
   "source": [
    "class LossHistory(keras.callbacks.Callback):\n",
    "    def on_train_begin(self, logs={}):\n",
    "        self.losses = {'batch':[], 'epoch':[]}\n",
    "        self.accuracy = {'batch':[], 'epoch':[]}\n",
    "        self.val_loss = {'batch':[], 'epoch':[]}\n",
    "        self.val_acc = {'batch':[], 'epoch':[]}\n",
    " \n",
    "    def on_batch_end(self, batch, logs={}):\n",
    "        self.losses['batch'].append(logs.get('loss'))\n",
    "        self.accuracy['batch'].append(logs.get('accuracy'))\n",
    "        self.val_loss['batch'].append(logs.get('val_loss'))\n",
    "        self.val_acc['batch'].append(logs.get('val_acc'))\n",
    " \n",
    "    def on_epoch_end(self, batch, logs={}):\n",
    "        self.losses['epoch'].append(logs.get('loss'))\n",
    "        self.accuracy['epoch'].append(logs.get('accuracy'))\n",
    "        self.val_loss['epoch'].append(logs.get('val_loss'))\n",
    "        self.val_acc['epoch'].append(logs.get('val_acc'))\n",
    " \n",
    "    def loss_plot(self, loss_type):\n",
    "        iters = range(len(self.losses[loss_type]))\n",
    "        plt.figure()\n",
    "        # acc\n",
    "        plt.plot(iters, self.accuracy[loss_type], 'r', label='train acc')\n",
    "        # loss\n",
    "        plt.plot(iters, self.losses[loss_type], 'g', label='train loss')\n",
    "        if loss_type == 'epoch':\n",
    "            # val_acc\n",
    "            plt.plot(iters, self.val_acc[loss_type], 'b', label='val acc')\n",
    "            # val_loss\n",
    "            plt.plot(iters, self.val_loss[loss_type], 'k', label='val loss')\n",
    "        plt.grid(True)\n",
    "        plt.xlabel(loss_type)\n",
    "        plt.ylabel('acc-loss')\n",
    "        plt.legend(loc=\"upper right\")\n",
    "        plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "97bf3f06",
   "metadata": {},
   "outputs": [],
   "source": [
    "history = LossHistory()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "6302d3be",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "WARNING:tensorflow:From C:\\Users\\29459\\anaconda3\\envs\\python37\\lib\\site-packages\\keras\\backend\\tensorflow_backend.py:422: The name tf.global_variables is deprecated. Please use tf.compat.v1.global_variables instead.\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "WARNING:tensorflow:From C:\\Users\\29459\\anaconda3\\envs\\python37\\lib\\site-packages\\keras\\backend\\tensorflow_backend.py:422: The name tf.global_variables is deprecated. Please use tf.compat.v1.global_variables instead.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Epoch 1/1\n",
      "60000/60000 [==============================] - 141s 2ms/step - loss: 0.1344 - accuracy: 0.9586\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "<keras.callbacks.callbacks.History at 0x17617ab1b08>"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#模型编译\n",
    "model.compile(optimizer='adam', loss='categorical_crossentropy', metrics=['accuracy'])\n",
    "\n",
    "#训练\n",
    "model.fit(X_train, y_train, batch_size=128, epochs=1,verbose=1,callbacks=[history])\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4288a80c",
   "metadata": {},
   "source": [
    "（4）利用准确率等指标评估所训练的神经网络；"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b37c0afa",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "ERROR:root:Internal Python error in the inspect module.\n",
      "Below is the traceback from this internal error.\n",
      "\n",
      "ERROR:root:Internal Python error in the inspect module.\n",
      "Below is the traceback from this internal error.\n",
      "\n",
      "ERROR:root:Internal Python error in the inspect module.\n",
      "Below is the traceback from this internal error.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\29459\\anaconda3\\envs\\python37\\lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 3437, in run_code\n",
      "    exec(code_obj, self.user_global_ns, self.user_ns)\n",
      "  File \"<ipython-input-7-172e1061fd60>\", line 1, in <module>\n",
      "    print(history.iters)\n",
      "AttributeError: 'LossHistory' object has no attribute 'iters'\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\29459\\anaconda3\\envs\\python37\\lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 2061, in showtraceback\n",
      "    stb = value._render_traceback_()\n",
      "AttributeError: 'AttributeError' object has no attribute '_render_traceback_'\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\29459\\anaconda3\\envs\\python37\\lib\\site-packages\\IPython\\core\\ultratb.py\", line 1101, in get_records\n",
      "    return _fixed_getinnerframes(etb, number_of_lines_of_context, tb_offset)\n",
      "  File \"C:\\Users\\29459\\anaconda3\\envs\\python37\\lib\\site-packages\\IPython\\core\\ultratb.py\", line 248, in wrapped\n",
      "    return f(*args, **kwargs)\n",
      "  File \"C:\\Users\\29459\\anaconda3\\envs\\python37\\lib\\site-packages\\IPython\\core\\ultratb.py\", line 281, in _fixed_getinnerframes\n",
      "    records = fix_frame_records_filenames(inspect.getinnerframes(etb, context))\n",
      "  File \"C:\\Users\\29459\\anaconda3\\envs\\python37\\lib\\inspect.py\", line 1502, in getinnerframes\n",
      "    frameinfo = (tb.tb_frame,) + getframeinfo(tb, context)\n",
      "  File \"C:\\Users\\29459\\anaconda3\\envs\\python37\\lib\\inspect.py\", line 1460, in getframeinfo\n",
      "    filename = getsourcefile(frame) or getfile(frame)\n",
      "  File \"C:\\Users\\29459\\anaconda3\\envs\\python37\\lib\\inspect.py\", line 696, in getsourcefile\n",
      "    if getattr(getmodule(object, filename), '__loader__', None) is not None:\n",
      "  File \"C:\\Users\\29459\\anaconda3\\envs\\python37\\lib\\inspect.py\", line 733, in getmodule\n",
      "    if ismodule(module) and hasattr(module, '__file__'):\n",
      "  File \"C:\\Users\\29459\\anaconda3\\envs\\python37\\lib\\site-packages\\tensorflow\\__init__.py\", line 50, in __getattr__\n",
      "    module = self._load()\n",
      "  File \"C:\\Users\\29459\\anaconda3\\envs\\python37\\lib\\site-packages\\tensorflow\\__init__.py\", line 44, in _load\n",
      "    module = _importlib.import_module(self.__name__)\n",
      "  File \"C:\\Users\\29459\\anaconda3\\envs\\python37\\lib\\importlib\\__init__.py\", line 127, in import_module\n",
      "    return _bootstrap._gcd_import(name[level:], package, level)\n",
      "  File \"<frozen importlib._bootstrap>\", line 1006, in _gcd_import\n",
      "  File \"<frozen importlib._bootstrap>\", line 983, in _find_and_load\n",
      "  File \"<frozen importlib._bootstrap>\", line 967, in _find_and_load_unlocked\n",
      "  File \"<frozen importlib._bootstrap>\", line 677, in _load_unlocked\n",
      "  File \"<frozen importlib._bootstrap_external>\", line 728, in exec_module\n",
      "  File \"<frozen importlib._bootstrap>\", line 219, in _call_with_frames_removed\n",
      "  File \"C:\\Users\\29459\\anaconda3\\envs\\python37\\lib\\site-packages\\tensorflow_core\\contrib\\__init__.py\", line 39, in <module>\n",
      "    from tensorflow.contrib import compiler\n",
      "  File \"C:\\Users\\29459\\anaconda3\\envs\\python37\\lib\\site-packages\\tensorflow_core\\contrib\\compiler\\__init__.py\", line 21, in <module>\n",
      "    from tensorflow.contrib.compiler import jit\n",
      "  File \"C:\\Users\\29459\\anaconda3\\envs\\python37\\lib\\site-packages\\tensorflow_core\\contrib\\compiler\\__init__.py\", line 22, in <module>\n",
      "    from tensorflow.contrib.compiler import xla\n",
      "  File \"C:\\Users\\29459\\anaconda3\\envs\\python37\\lib\\site-packages\\tensorflow_core\\contrib\\compiler\\xla.py\", line 22, in <module>\n",
      "    from tensorflow.python.estimator import model_fn as model_fn_lib\n",
      "  File \"C:\\Users\\29459\\anaconda3\\envs\\python37\\lib\\site-packages\\tensorflow_core\\python\\estimator\\model_fn.py\", line 26, in <module>\n",
      "    from tensorflow_estimator.python.estimator import model_fn\n",
      "ModuleNotFoundError: No module named 'tensorflow_estimator.python'\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\29459\\anaconda3\\envs\\python37\\lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 3437, in run_code\n",
      "    exec(code_obj, self.user_global_ns, self.user_ns)\n",
      "  File \"<ipython-input-7-172e1061fd60>\", line 1, in <module>\n",
      "    print(history.iters)\n",
      "AttributeError: 'LossHistory' object has no attribute 'iters'\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\29459\\anaconda3\\envs\\python37\\lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 2061, in showtraceback\n",
      "    stb = value._render_traceback_()\n",
      "AttributeError: 'AttributeError' object has no attribute '_render_traceback_'\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\29459\\anaconda3\\envs\\python37\\lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 3357, in run_ast_nodes\n",
      "    if (await self.run_code(code, result,  async_=asy)):\n",
      "  File \"C:\\Users\\29459\\anaconda3\\envs\\python37\\lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 3454, in run_code\n",
      "    self.showtraceback(running_compiled_code=True)\n",
      "  File \"C:\\Users\\29459\\anaconda3\\envs\\python37\\lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 2064, in showtraceback\n",
      "    value, tb, tb_offset=tb_offset)\n",
      "  File \"C:\\Users\\29459\\anaconda3\\envs\\python37\\lib\\site-packages\\IPython\\core\\ultratb.py\", line 1368, in structured_traceback\n",
      "    self, etype, value, tb, tb_offset, number_of_lines_of_context)\n",
      "  File \"C:\\Users\\29459\\anaconda3\\envs\\python37\\lib\\site-packages\\IPython\\core\\ultratb.py\", line 1268, in structured_traceback\n",
      "    self, etype, value, tb, tb_offset, number_of_lines_of_context\n",
      "  File \"C:\\Users\\29459\\anaconda3\\envs\\python37\\lib\\site-packages\\IPython\\core\\ultratb.py\", line 1125, in structured_traceback\n",
      "    tb_offset)\n",
      "  File \"C:\\Users\\29459\\anaconda3\\envs\\python37\\lib\\site-packages\\IPython\\core\\ultratb.py\", line 1082, in format_exception_as_a_whole\n",
      "    last_unique, recursion_repeat = find_recursion(orig_etype, evalue, records)\n",
      "  File \"C:\\Users\\29459\\anaconda3\\envs\\python37\\lib\\site-packages\\IPython\\core\\ultratb.py\", line 382, in find_recursion\n",
      "    return len(records), 0\n",
      "TypeError: object of type 'NoneType' has no len()\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\29459\\anaconda3\\envs\\python37\\lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 2061, in showtraceback\n",
      "    stb = value._render_traceback_()\n",
      "AttributeError: 'TypeError' object has no attribute '_render_traceback_'\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\29459\\anaconda3\\envs\\python37\\lib\\site-packages\\IPython\\core\\ultratb.py\", line 1101, in get_records\n",
      "    return _fixed_getinnerframes(etb, number_of_lines_of_context, tb_offset)\n",
      "  File \"C:\\Users\\29459\\anaconda3\\envs\\python37\\lib\\site-packages\\IPython\\core\\ultratb.py\", line 248, in wrapped\n",
      "    return f(*args, **kwargs)\n",
      "  File \"C:\\Users\\29459\\anaconda3\\envs\\python37\\lib\\site-packages\\IPython\\core\\ultratb.py\", line 281, in _fixed_getinnerframes\n",
      "    records = fix_frame_records_filenames(inspect.getinnerframes(etb, context))\n",
      "  File \"C:\\Users\\29459\\anaconda3\\envs\\python37\\lib\\inspect.py\", line 1502, in getinnerframes\n",
      "    frameinfo = (tb.tb_frame,) + getframeinfo(tb, context)\n",
      "  File \"C:\\Users\\29459\\anaconda3\\envs\\python37\\lib\\inspect.py\", line 1460, in getframeinfo\n",
      "    filename = getsourcefile(frame) or getfile(frame)\n",
      "  File \"C:\\Users\\29459\\anaconda3\\envs\\python37\\lib\\inspect.py\", line 696, in getsourcefile\n",
      "    if getattr(getmodule(object, filename), '__loader__', None) is not None:\n",
      "  File \"C:\\Users\\29459\\anaconda3\\envs\\python37\\lib\\inspect.py\", line 733, in getmodule\n",
      "    if ismodule(module) and hasattr(module, '__file__'):\n",
      "  File \"C:\\Users\\29459\\anaconda3\\envs\\python37\\lib\\site-packages\\tensorflow\\__init__.py\", line 50, in __getattr__\n",
      "    module = self._load()\n",
      "  File \"C:\\Users\\29459\\anaconda3\\envs\\python37\\lib\\site-packages\\tensorflow\\__init__.py\", line 44, in _load\n",
      "    module = _importlib.import_module(self.__name__)\n",
      "  File \"C:\\Users\\29459\\anaconda3\\envs\\python37\\lib\\importlib\\__init__.py\", line 127, in import_module\n",
      "    return _bootstrap._gcd_import(name[level:], package, level)\n",
      "  File \"<frozen importlib._bootstrap>\", line 1006, in _gcd_import\n",
      "  File \"<frozen importlib._bootstrap>\", line 983, in _find_and_load\n",
      "  File \"<frozen importlib._bootstrap>\", line 965, in _find_and_load_unlocked\n",
      "ModuleNotFoundError: No module named 'tensorflow_core.estimator'\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\29459\\anaconda3\\envs\\python37\\lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 3437, in run_code\n",
      "    exec(code_obj, self.user_global_ns, self.user_ns)\n",
      "  File \"<ipython-input-7-172e1061fd60>\", line 1, in <module>\n",
      "    print(history.iters)\n",
      "AttributeError: 'LossHistory' object has no attribute 'iters'\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\29459\\anaconda3\\envs\\python37\\lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 2061, in showtraceback\n",
      "    stb = value._render_traceback_()\n",
      "AttributeError: 'AttributeError' object has no attribute '_render_traceback_'\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\29459\\anaconda3\\envs\\python37\\lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 3357, in run_ast_nodes\n",
      "    if (await self.run_code(code, result,  async_=asy)):\n",
      "  File \"C:\\Users\\29459\\anaconda3\\envs\\python37\\lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 3454, in run_code\n",
      "    self.showtraceback(running_compiled_code=True)\n",
      "  File \"C:\\Users\\29459\\anaconda3\\envs\\python37\\lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 2064, in showtraceback\n",
      "    value, tb, tb_offset=tb_offset)\n",
      "  File \"C:\\Users\\29459\\anaconda3\\envs\\python37\\lib\\site-packages\\IPython\\core\\ultratb.py\", line 1368, in structured_traceback\n",
      "    self, etype, value, tb, tb_offset, number_of_lines_of_context)\n",
      "  File \"C:\\Users\\29459\\anaconda3\\envs\\python37\\lib\\site-packages\\IPython\\core\\ultratb.py\", line 1268, in structured_traceback\n",
      "    self, etype, value, tb, tb_offset, number_of_lines_of_context\n",
      "  File \"C:\\Users\\29459\\anaconda3\\envs\\python37\\lib\\site-packages\\IPython\\core\\ultratb.py\", line 1125, in structured_traceback\n",
      "    tb_offset)\n",
      "  File \"C:\\Users\\29459\\anaconda3\\envs\\python37\\lib\\site-packages\\IPython\\core\\ultratb.py\", line 1082, in format_exception_as_a_whole\n",
      "    last_unique, recursion_repeat = find_recursion(orig_etype, evalue, records)\n",
      "  File \"C:\\Users\\29459\\anaconda3\\envs\\python37\\lib\\site-packages\\IPython\\core\\ultratb.py\", line 382, in find_recursion\n",
      "    return len(records), 0\n",
      "TypeError: object of type 'NoneType' has no len()\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\29459\\anaconda3\\envs\\python37\\lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 2061, in showtraceback\n",
      "    stb = value._render_traceback_()\n",
      "AttributeError: 'TypeError' object has no attribute '_render_traceback_'\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\29459\\anaconda3\\envs\\python37\\lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 2940, in _run_cell\n",
      "    return runner(coro)\n",
      "  File \"C:\\Users\\29459\\anaconda3\\envs\\python37\\lib\\site-packages\\IPython\\core\\async_helpers.py\", line 68, in _pseudo_sync_runner\n",
      "    coro.send(None)\n",
      "  File \"C:\\Users\\29459\\anaconda3\\envs\\python37\\lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 3166, in run_cell_async\n",
      "    interactivity=interactivity, compiler=compiler, result=result)\n",
      "  File \"C:\\Users\\29459\\anaconda3\\envs\\python37\\lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 3376, in run_ast_nodes\n",
      "    self.showtraceback()\n",
      "  File \"C:\\Users\\29459\\anaconda3\\envs\\python37\\lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 2064, in showtraceback\n",
      "    value, tb, tb_offset=tb_offset)\n",
      "  File \"C:\\Users\\29459\\anaconda3\\envs\\python37\\lib\\site-packages\\IPython\\core\\ultratb.py\", line 1368, in structured_traceback\n",
      "    self, etype, value, tb, tb_offset, number_of_lines_of_context)\n",
      "  File \"C:\\Users\\29459\\anaconda3\\envs\\python37\\lib\\site-packages\\IPython\\core\\ultratb.py\", line 1268, in structured_traceback\n",
      "    self, etype, value, tb, tb_offset, number_of_lines_of_context\n",
      "  File \"C:\\Users\\29459\\anaconda3\\envs\\python37\\lib\\site-packages\\IPython\\core\\ultratb.py\", line 1143, in structured_traceback\n",
      "    chained_exceptions_tb_offset)\n",
      "  File \"C:\\Users\\29459\\anaconda3\\envs\\python37\\lib\\site-packages\\IPython\\core\\ultratb.py\", line 1082, in format_exception_as_a_whole\n",
      "    last_unique, recursion_repeat = find_recursion(orig_etype, evalue, records)\n",
      "  File \"C:\\Users\\29459\\anaconda3\\envs\\python37\\lib\\site-packages\\IPython\\core\\ultratb.py\", line 382, in find_recursion\n",
      "    return len(records), 0\n",
      "TypeError: object of type 'NoneType' has no len()\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\29459\\anaconda3\\envs\\python37\\lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 2061, in showtraceback\n",
      "    stb = value._render_traceback_()\n",
      "AttributeError: 'TypeError' object has no attribute '_render_traceback_'\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\29459\\anaconda3\\envs\\python37\\lib\\site-packages\\IPython\\core\\ultratb.py\", line 1101, in get_records\n",
      "    return _fixed_getinnerframes(etb, number_of_lines_of_context, tb_offset)\n",
      "  File \"C:\\Users\\29459\\anaconda3\\envs\\python37\\lib\\site-packages\\IPython\\core\\ultratb.py\", line 248, in wrapped\n",
      "    return f(*args, **kwargs)\n",
      "  File \"C:\\Users\\29459\\anaconda3\\envs\\python37\\lib\\site-packages\\IPython\\core\\ultratb.py\", line 281, in _fixed_getinnerframes\n",
      "    records = fix_frame_records_filenames(inspect.getinnerframes(etb, context))\n",
      "  File \"C:\\Users\\29459\\anaconda3\\envs\\python37\\lib\\inspect.py\", line 1502, in getinnerframes\n",
      "    frameinfo = (tb.tb_frame,) + getframeinfo(tb, context)\n",
      "  File \"C:\\Users\\29459\\anaconda3\\envs\\python37\\lib\\inspect.py\", line 1460, in getframeinfo\n",
      "    filename = getsourcefile(frame) or getfile(frame)\n",
      "  File \"C:\\Users\\29459\\anaconda3\\envs\\python37\\lib\\inspect.py\", line 696, in getsourcefile\n",
      "    if getattr(getmodule(object, filename), '__loader__', None) is not None:\n",
      "  File \"C:\\Users\\29459\\anaconda3\\envs\\python37\\lib\\inspect.py\", line 733, in getmodule\n",
      "    if ismodule(module) and hasattr(module, '__file__'):\n",
      "  File \"C:\\Users\\29459\\anaconda3\\envs\\python37\\lib\\site-packages\\tensorflow\\__init__.py\", line 50, in __getattr__\n",
      "    module = self._load()\n",
      "  File \"C:\\Users\\29459\\anaconda3\\envs\\python37\\lib\\site-packages\\tensorflow\\__init__.py\", line 44, in _load\n",
      "    module = _importlib.import_module(self.__name__)\n",
      "  File \"C:\\Users\\29459\\anaconda3\\envs\\python37\\lib\\importlib\\__init__.py\", line 127, in import_module\n",
      "    return _bootstrap._gcd_import(name[level:], package, level)\n",
      "  File \"<frozen importlib._bootstrap>\", line 1006, in _gcd_import\n",
      "  File \"<frozen importlib._bootstrap>\", line 983, in _find_and_load\n",
      "  File \"<frozen importlib._bootstrap>\", line 965, in _find_and_load_unlocked\n",
      "ModuleNotFoundError: No module named 'tensorflow_core.estimator'\n"
     ]
    }
   ],
   "source": [
    "print(history.iters)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "a3fbf7b7",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ERROR! Session/line number was not unique in database. History logging moved to new session 295\n"
     ]
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAYIAAAEGCAYAAABo25JHAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjMuNCwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy8QVMy6AAAACXBIWXMAAAsTAAALEwEAmpwYAAA8eElEQVR4nO3dd3hUVf748fdJJ4UACYReBQEFgnRRmq4Ud7GsigoKrop+7QV37aL+XHsFXURBRQTsKIrYSKgizdA7ARJaSEjv5fz+uDM3M5lJGEImk3A/r+eZZ26/Zw7hfu4p91yltUYIIYR1+fk6AUIIIXxLAoEQQlicBAIhhLA4CQRCCGFxEgiEEMLiAnydgNMVHR2t27dvX619c3NzCQsLq9kE1TOSB5IHdpIP1sqDDRs2pGqtm7pbV+8CQfv27Vm/fn219o2Pj2fYsGE1m6B6RvJA8sBO8sFaeaCUOljZOqkaEkIIi5NAIIQQFieBQAghLK7etREIIc5excXFJCcnU1BQUCvni4yMZMeOHbVyrtoSEhJC69atCQwM9HgfCQRCiDojOTmZiIgI2rdvj1LK6+fLzs4mIiLC6+epLVpr0tLSSE5OpkOHDh7vJ1VDQog6o6CggKioqFoJAmcjpRRRUVGnXaKSQCCEqFMkCJyZ6uSfZQLBluNbmJU4i7S8NF8nRQgh6hTLBII9J/cw99BckrKSfJ0UIUQdlZGRwXvvvVetfceMGUNGRkbNJqiWWCYQRDWIApASgRCiUlUFgtLS0ir3Xbx4MY0aNfJCqrzPOoEg1BYI8iUQCCHce/TRR9m3bx+xsbE88sgjxMfHM3z4cG688UZ69OgBwJVXXkmfPn0477zzmDlzprlv+/btSU1N5cCBA3Tr1o3bb7+d8847j8suu4z8/HyXcy1atIgBAwbQu3dvLr30Uo4fPw5ATk4Ot9xyCz169KBnz558/fXXACxZsoQLLriAXr16cckll9To77ZM99Ho0GhASgRC1BsPPAAJCTV7zNhYeOutSle/9NJLbN26lQTbeePj41m7di1bt241u2POnj2bJk2akJ+fT79+/fjnP/9JVFSU03H27NnD/Pnz+eCDD7juuuv4+uuvmTBhgtM2F110EWvWrEEpxYcffsgrr7zC66+/zvPPP09kZCRbtmwBID09nRMnTnD77bezfPlyOnTowMmTJ2ssS8BCgaBJgyaAlAiEEKenf//+Tn3y33nnHb799lsAkpKS2LNnj0sg6NChA7GxsQD06dOHAwcOuBw3OTmZcePGcfToUYqKisxz/PbbbyxYsMDcrnHjxixatIghQ4aY2zRp0qQmf6J1AkGQfxCh/qGk5qX6OilCCE9UcedemxyHqY6Pj+e3337jjz/+IDQ0lGHDhrntsx8cHGxO+/v7u60auvfee3nooYcYO3Ys8fHxTJ06FTAeCqvYBdTdsppkmTYCgIaBDaVEIISoVEREBNnZ2ZWuz8zMpHHjxoSGhrJz507WrFlT7XNlZmbSqlUrAD755BNz+WWXXcb06dPN+fT0dAYNGsSyZctITEwEqPGqIWsFgoCG0kYghKhUVFQUgwcP5vzzz+eRRx5xWT9q1ChKSkro2bMnTz31FAMHDqz2uaZOncq1117LxRdfTHR0tLn8ySefJD09nfPPP59evXoRFxdH06ZNmTlzJldffTW9evVi3Lhx1T6vO0prXaMH9La+ffvq6r6Ypt87/fAL9ePP2/6s4VTVH1Z6EUdlJA8MdTEfduzYQbdu3WrtfGfbWEN27vJRKbVBa93X3faWKhFEBESQUZDh62QIIUSdYqlAEBYQRmZBpq+TIYQQdYq1AoF/GFmFWb5OhhBC1CmWCgShAaHkl+RTXFrs66QIIUSdYalAEOZv9AeWUoEQQpSzViAIkEAghBAVWTIQZBZKg7EQwlVtDkM9depUXnvttWqdq6ZZKhCE+ocCUiIQQrgnw1BbQHhAOIB0IRVCuFWbw1A7SkhIYODAgfTs2ZOrrrqK9PR0wBjgrnv37vTs2ZPrr78egGXLlhEbG0tsbCy9e/euckgMT1lm0DmQEoEQ9ckDSx4g4VhCjR4ztnksb416q9L1tTkMtaObb76ZadOmMXToUJ5++mmeffZZ3nrrLV566SUSExMJDg42q51ee+013n33XQYPHkxOTg4hISFnlCdgsRKBtBEIIU6Xu2Goe/XqxcCBA81hqCvyZBhqu8zMTDIyMhg6dCgAEydOZPny5QD07NmT8ePHM3fuXAICjPv2wYMH89BDD/HOO++QkZFhLj8TUiIQQtRJVd251yZvDUPtiR9//JHly5fz/fff8/zzz7Nt2zYeffRRLr/8chYvXszAgQP57bff6Nq1a7WOb2epEkGwn/GPk1ec5+OUCCHqotochtouMjKSxo0bs2LFCgA+/fRThg4dSllZGUlJSQwfPpxXXnmFjIwMcnJy2LdvHz169OA///kPffv2ZefOnWecBkuVCJRShAaGSiAQQrjlOAz16NGjufzyy53Wjxo1ihkzZtCzZ0/OPffcMxqG2tEnn3zCnXfeSV5eHh07duSjjz6itLSUCRMmkJmZidaaBx98kEaNGvHUU08RFxeHv78/3bt3Z/To0Wd8fq8FAqVUG2AO0BwoA2Zqrd+usI0C3gbGAHnAJK31Rm+lCZBAIISo0rx585zmHYfqDg4O5qeffnK7n70dIDo6mq1bt5rLp0yZ4nZ7+xvJAGJjY92WLlauXOmybNq0aZUlvdq8WSIoAR7WWm9USkUAG5RSv2qttztsMxrobPsMAP5n+/YaCQRCCOHMa20EWuuj9rt7rXU2sANoVWGzK4A52rAGaKSUauGtNIEEAiGEqKhW2giUUu2B3kDFV4O1ApIc5pNty45W2H8yMBkgJiaG+Pj4aqUjJyeH0oJSko4lVfsY9V1OTo5lf7ud5IGhLuZDZGQkWVlZXn1Ru6PS0tIaeSCrLtFaU1BQcFr/tl4PBEqpcOBr4AGtdcV+m+7+tV3enam1ngnMBONVldV9vV58fDwxTWII9Ausc6/oqy118fWEtU3ywFAX8yExMZGioiKioqJqJRicba+q1FqTlpZGo0aN6N27t8f7eTUQKKUCMYLAZ1rrb9xskgy0cZhvDRzxZppCA0NliAkh6qjWrVuTnJzMiRMnauV8BQUFNfJkbl0SEhJC69atT2sfb/YaUsAsYIfW+o1KNvseuEcptQCjkThTa320km1rRGhgKEezvXoKIUQ1BQYGOj3F623x8fGnded8tvJmiWAwcBOwRSmVYFv2ONAWQGs9A1iM0XV0L0b30Vu8mB5AGouFEKIirwUCrfVK3LcBOG6jgbu9lQZ3QgMkEAghhCNLDTEBUiIQQoiKJBAIIYTFWTIQFJcVU1xa7OukCCFEnWDJQACQX1K9YWGFEOJsY9lAINVDQghhkEAghBAWJ4FACCEsTgKBEEJYnAQCIYSwOAkEQghhcZYNBPnF0n1UCCHAwoFASgRCCGGQQCCEEBYngUAIISxOAoEQQlic5QJBoH8gAX4BEgiEEMLGcoEAZChqIYRwJIFACCEszrqBoEQCgRBCgJUDgZQIhBACkEAghBCWZ9lAkFuU6+tkCCFEnWDJQBAWGEZOUY6vkyGEEHWCJQNBw+CGZBdl+zoZQghRJ1gyEEQERZBVmOXrZAghRJ1gyUDQMLgh2YVSIhBCCLBoIIgIjiC/JJ+SshJfJ0UIIXzOkoGgYXBDACkVCCEEFg0EEUERANJgLIQQWDUQBNsCgZQIhBDCmoHAXjUkPYeEEMKigUCqhoQQopwlA4E0FgshRDlLBgJ7G4FUDQkhhBcDgVJqtlIqRSm1tZL1w5RSmUqpBNvnaW+lpaLwoHAAcotl4DkhhAjw4rE/BqYDc6rYZoXW+u9eTINbQf5BABSVFtX2qYUQos7xWolAa70cOOmt458JeyAoLi32cUqEEML3vFki8MQgpdQm4AgwRWu9zd1GSqnJwGSAmJgY4uPjq3WynJwc4uPjKdWlAOzat4v4kuodq76y54GVSR4YJB8kD+x8GQg2Au201jlKqTHAQqCzuw211jOBmQB9+/bVw4YNq9YJ4+Pjse/rt8KPVm1aUd1j1VeOeWBVkgcGyQfJAzuf9RrSWmdprXNs04uBQKVUdG2dP8g/SNoIhBACHwYCpVRzpZSyTfe3pSWtts4f6BdIcZm0EQghhNeqhpRS84FhQLRSKhl4BggE0FrPAK4B/k8pVQLkA9drrbW30lORlAiEEMLgtUCgtb7hFOunY3Qv9QkJBEIIYbDkk8UAgf5SNSSEEGDhQCAlAiGEMEggEEIIi5NAIIQQFmfZQBDoFyhDTAghBBYOBFIiEEIIgwQCIYSwOI8CgVJqsFIqzDY9QSn1hlKqnXeT5l2B/oESCIQQAs9LBP8D8pRSvYB/Awep+j0DdV6Qf5A8RyCEEHgeCEpswz9cAbyttX4biPBesrxPqoaEEMLg6RAT2Uqpx4AJwBCllD+2cYPqq0A/qRoSQgjwvEQwDigEbtVaHwNaAa96LVW1IMg/SLqPCiEEp1EiwKgSKlVKdQG6AvO9lyzvk6ohIYQweFoiWA4EK6VaAb8Dt2C8nL7ekqohIYQweBoIlNY6D7gamKa1vgo4z3vJ8j4pEQghhMHjQKCUGgSMB360LfP3TpJqh3QfFUIIg6eB4AHgMeBbrfU2pVRHIM5rqaoF8kCZEEIYPGos1lovA5YppSKUUuFa6/3Afd5NmncF+QdRUlZCmS7DT1l2pA0hhPB4iIkeSqm/gK3AdqXUBqVUvW8jAKQLqRDC8jy9FX4feEhr3U5r3RZ4GPjAe8nyvkA/43k4aScQQlidp4EgTGtttgloreOBMK+kqJaEBIQAkF+c7+OUCCGEb3kaCPYrpZ5SSrW3fZ4EEr2ZMG9rFtYMgOO5x32cEiGE8C1PA8G/gKbAN8C3tulbvJWo2tAiogUAR7OP+jglQgjhW572GkqnnvcSqqhFuC0Q5EggEEJYW5WBQCm1CNCVrddaj63xFNUSKREIIYThVCWC12olFT4QHhROeFC4lAiEEJZXZSCwPUjmRCl1gdZ6o/eSVHtahLfgWM4xXydDCCF8qjqP1H5Y46nwkejQaNLy03ydDCGE8KnqBAJV46nwkeCAYApLCn2dDCGE8KnqBIJnazwVPhLsH0xhqQQCIYS1eTrW0FVKqUgArfVCpVQjpdSVXk1ZLZASgRBCeF4ieEZrnWmf0VpnAM94JUW1KCQgREoEQgjL8zQQuNvO0/cd11nB/lIiEEIITwPBeqXUG0qpTkqpjkqpN4EN3kxYbQj2D6agpMDXyRBCCJ/yNBDcCxQBnwNfAPnA3VXtoJSarZRKUUptrWS9Ukq9o5Taq5TarJS64HQSXhOCA6SxWAghPB1rKBd49DSP/TEwHZhTyfrRQGfbZwDwP9t3rZGqISGE8LzX0K9KqUYO842VUj9XtY/WejlwsopNrgDmaMMaoJFSqoUn6akpUiIQQgjPq4aibT2FAHM00mZneO5WQJLDfLJtWa0J9g8231sshBBW5WnPnzKlVFut9SEApVR7qhiV1EPunlB2e0yl1GRgMkBMTAzx8fHVOmFOTo7TvkcOHQHg16W/EuwfXK1j1jcV88CKJA8Mkg+SB3aeBoIngJVKKfsgdEOwXZjPQDLQxmG+NXDE3YZa65nATIC+ffvqYcOGVeuE8fHxOO6bsCYBEqH/hf1p3KBxtY5Z31TMAyuSPDBIPkge2HnaWLxEKdUX4+KfAHyH0XPoTHwP3KOUWoDRSJypta7VMaHtpQBpJxDCx8rKoLTU+Librurbk20q+Y5KSICMjBo5ltt02T9aVz3t6fqxY+GGG2o8+z0KBEqp24D7Me7aE4CBwB/AiCr2mQ8MA6KVUskYTyIHAmitZwCLgTHAXiAPH7z6MjjAFgik55CoC+z/2YuLoaTE9dvdsjP4brdnD8TF1fw57Bdxx4/jxd3dMh/p4a0D+/kZH39/UKp83j7tbpkn6/v29UpyPa0auh/oB6zRWg9XSnXlFIPPaa2rDFtaa80pnkXwNikRWEBZGRQUQH6+8V1QQOjBg5CQAIWFxqeoqHza3XxV2xQV1ejFuTZ1sE/4+0NgIAQEnN53cDCEhzsvDwgwjufuY78wnu4y+3xV355s4+Z7/caN9B0w4IyO4fKt6t8AzZ4GggKtdYFSCqVUsNZ6p1LqXK+mrBZIiaAOKC2F3FzIzoacHPef3Fz3n7w84wLv7mNfV+j6b9u/umkNCjIufsHB5dNBQcZFsOKFMiTEuEhW5wJrv6BWd18Pv5etXs3QESPq5YWrpuTk5kJsrK+T4XOeBoJk23MEC4FflVLpVNKwW59IieAMFRdDerrzJyMDMjPLvx0/WVnGBd/+nZNjXKw9pRSEhZV/QkONT4MGEBVlfJ/qExzM9n376N67t+tFvbL54GDj4nmWXTB1Pb17FTXP08biq2yTU5VScUAksMRrqaol9hKBjDeEUcVx4gSkpDh/26dTU10v+rm5VR8zMBAiI8s/DRtC27YQEVH+CQ+v+uN44Q8JqZELV0p8PN2lp4gQptMeQdTde4zrq5CAEOAsrRoqKXF7Ue+wbh3Mm+e6LjPT/XECAqBpU+OOu0kT6NgRGjcu/zRp4jzfqJFx0W/UqMYu3EII76r3Q0mfiXpdNVRQAAcOwP79xmffPmM+ORkOH4bjx42G0gra+vlBs2bGxb1ZM6MXgn3a3XejRnIxF+IsZ+1AUB8ai3NzYe9e2LYNtm4t/xw4YHQ3tAsNhfbtoXVr6NHD+G7RAmJijAu67eK+LCGBYSMq7fUrhLAgaweCulYiOHoUNmyA9euN77/+Mu7u7QIC4NxzoX9/mDgROnUyqmo6djQu+J7cuftV5zXVQoizmaUDgb2NoKCkgMKSQrOEUCtKSmDTJli5ElasgNWrjUAAxsW6a1cYMcL47tQJunc3gkBQUO2lUQhhCZYOBOFB4QCsO7yOW767hS+v/ZJrul/jvRNu3w7ffAPLl8MffxjdJ8Go0hkxAvr1M+rsY2ONXjJCCFELJBAAyw4aHaG+3/V9zQeCPXvg88+Nz9atRvVNz55G1c7FF8PgwUZ9vhBC+IilA0FIQAh+yo/somwA/FQN1Z8XFsJnn8G778LGjcayiy6CadPgmmugefOaOY8QQtQASwcCpRThQeFkFWaZ82ckMRFefx0+/tjo7dOrlzF/7bXQps0pdxdCCF+wdCAAo3roWM4xAPw8fmFbBVu2wMsvw4IFRkPv+PHG55JLpA++EKLOs3xfwvCgcKdXVc7dPJeSshLPdk5NhdtuM+r8Fy6EBx4wSgUffQSXXipBQAhRL0iJwNZgDDBn8xxmJ8wmJTeFhwY9VPWOP/4IN99sDKA2ZQo89pgx3IIQQtQzUiJwCAT2ksDxnOOV71BaCs88A3//O7RrZ4xr/+qrEgSEEPWWlAgcAoGdv5+/+43T0+HGG2HJEpg0Cd57zxjaWAgh6jEJBG4CgdtupGlpRr3/tm3w/vtw++3SBiCEOCtIIAh0UyJQFUoEaWlGD6CdO2HRIhg5spZSJ4QQ3ieB4FQlgsJCuOoqIwh8/z1cdlktpk4IIbxPAkFVgUBruOMOY1C4+fMlCAghzkqW7zUUGRLpsqyotMiYmDULPvnE6CV0/fW1nDIhhKgdlg8ErRu6DvhWUFJgDAk9ZQoMGwZPP137CRNCiFpi+UDQNrKty7L8kny4917jdZAzZ8rLXIQQZzXLtxG0aeg6GFzBvl3w9W/w4ovQubMPUiWEELXH8re6LSNauizLX7vKGDn04Yd9kCIhhKhdlg8Egf6BLssKivJhxgwIdF0nhBBnG8sHAoA3R75Jg4DyoSLyO7SCgQN9mCIhhKg9EgiABwY+wPiu15nzBed08GFqhBCidkkgsNuy2ZzMVx6+j0AIIc4CEggAior48+QWwHiPcUFJgY8TJIQQtUcCAcAvv/DSkhKuaTSYv3f5uwQCIYSlSCAA+PxzxqQ14cu74ggLDCO/ON/XKRJCiFojgaCgAL77Dq6+GgIDCQkIIbc4F621r1MmhBC1wquBQCk1Sim1Sym1Vyn1qJv1w5RSmUqpBNun9gf1WbIEsrPhOqPXUGzzWFLzUll3ZF2tJ0UIIXzBa4FAKeUPvAuMBroDNyilurvZdIXWOtb2ec5b6anUkiXQsCEMHw7AjT1upEFAAz7b/FmtJ0UIIXzBmyWC/sBerfV+rXURsAC4wovnq57ly+GiiyDAGHapYXBDesT0YNuJbT5OmBBC1A5vBoJWQJLDfLJtWUWDlFKblFI/KaXO82J6XKWkwI4dMGSI0+Ku0V3ZmbqzVpMihBC+4s3RR9292b1iC+xGoJ3WOkcpNQZYCLgM96mUmgxMBoiJiSE+Pr5aCcrJyXHaN3r5cs4HNoaHk+WwPCgriMPZh1n822I0mhJdQmSg6wts6qOKeWBFkgcGyQfJA5PW2isfYBDws8P8Y8Bjp9jnABBd1TZ9+vTR1RUXF+e84P77tW7QQOvCQqfF32z/RjMV/UfSH7rpK001U6n2OesalzywIMkDg+SDtfIAWK8rua56s2poHdBZKdVBKRUEXA9877iBUqq5UkrZpvtjVFWleTFNzpYvh0GDICjIafHA1saAcysOruBE3olaS44QQviC1wKB1roEuAf4GdgBfKG13qaUulMpdadts2uArUqpTcA7wPW2yOV9mZmQkAAXX+yyqkVEC7pGdyXuQJzbXct0GblFuV5OoBBC1A6vvqFMa70YWFxh2QyH6enAdG+moVKrVoHWLg3FdgNbD+SXfb+Y82W6DD9lxM1///pvXv/jdQqeKCA4ILhWkiuEEN5i3SeL1683vgcMcLs6JiyG1LxUc96xBPBRwkcAZBZmei99QghRS6wbCHbvhrZtISzM7ero0GiKSovM+azCLHM6yN9oU8gskEAghKj/rBsIdu2Cc8+tdHXT0KZO8453/8H+wS7LhBCivrJmINDaCARdulS6SXRotNP819u/5qc9PwGY7QIZBRleS6IQQtQWrzYW11k7dhgDzXV3N/SRoWIgeDreGA9PP6PLSwRSNSSEOAtYs0QwcyYEBsI111S6ScVA4MjeRiAlAiHE2cCagWDlShg2DJo1q3QTd4EgwC8ArbVZNSRtBEKIs4E1A0FiInTqVOUmDYMbuiwrKSshszCTAD+jRk1KBEKIs4H1AkFWFpw8CR06VLmZbeQLAMICw7it920ApOSmUFhSCEgbgRDi7GC9QJCYaHyfIhCA8VAZwMLrF3LdecYbzFJyU8yX22cUZvDcsuf4evvXTvtlF2aTlld7QyYJIcSZkEBQhTaRbQAI9AukaZjxXEFKbgr5JcbL7dPz03km/hmu+dK50bnztM5Ev1p5Y3NJWQkvr3yZ/OL86vwCIYSoUdYLBEm2d+W0a3fKTcd2GQtAeFA4zcKMhuWU3BTzAn4897jb/Spbbjf7r9k8+vujvLzqZU9TLYQQXmO95wiOHDG6jkZXfsdu98SQJxjTeQx9WvYxh5twLBHsSt1VrSTYq43sVUxCCOFL1isRHD4MLVuCcvcCNWd+yo8+LfsAxrMDjUMaO7UROHYfTc1L5Yavb/CoJ5F9/5dXvcwTvz9RjR8hhBA1x3qB4MgRIxBUQ7OwZuxK20VOUY7Luml/TmPB1gVMWjjJXFZaVur2OI4lgf+u/G+10iKEEDVFAsFpaBbWjN/2/wZA8/DmTuv+PPwnAN/t+s5clltcPnR1VmGW/XWcLlVCqw6tIiU3pVppqm35xflO72kQQtR/1gwELVpUa1fHp41bN2zttO7nfT+7bG8vOSRnJRP5UiTv/PkO64+sN9sY7C766CIGfjjQZf8tx7fwyC+PUNMvbfsj6Q/S89Orte9di+9i5NyR7Dixo0bT5C2rk1azNWWrr5MhRJ1mqcZi//x84xWV1SwRON7Je9LQa3+ZzcGMgwA89MtDlOkyt9smZhjdWotKi8xhLC6bexnHco7x78H/NruveuJk/kliZ8Ty5bVfMqC184t3SspKGD57OANaDWDNbWs8PqbdpmObjN9WXD9e1Tl49mDAGCxQCOGepUoEIceOGRPt21drf8fqmx7Nepxye3uJwB40KgsCji54/wI6T+sMQF5xHnD6F93lB5eTlJXEs8uedVmXXZINlFdlna7ismIA87WdQoj6z1L/m0OOHjUmPHiYzJ1JsZMA+PLaL5k1dhZ7793LvKvnVbr93M1zKSgpYOWhlR4dv6i0iG0ntpGUlYTW2mxs7vB2B37Y/YO5XXZhtjmdVZhlbrf+yHrWH1lvBiB3gSerOMtl2ekoLjUCgXR9FeLsIYHgNNzd726Knyrmmu7X0CCwAZ2adHLpQXR1t6vN6TfWvEGDFxowddlUj46/9vBaczo5K5lSXd7r6OOEjwFYk7yGhi81ZP6W+RSWFNLx7Y78b/3/AOj3QT/6fdDPrIoqKStxOUdWyRkGAluJwF5aEZVzfL2psK7XV79Ok5eb+DoZVbJWIDh2DEJDqxx+uipKKXPkUbsh7YY4zb858k26RFX+5rOqbD+x3ZyeuWGm0113gF8AczbNYdCsQYDR3rD9xHbS8tPYfHyz0/uVD2YagcDdE86nKhHsPbmXxPREkjKTzGU7U3cSfyAeKA8u+cX5FJQUcP1X17Mnbc9p/tKz39LEpUS+FMnSxKW+TorwsSm/TiG9IN3tjVldYa1AcPy4MbSEBw+Teerc6HPRz2geGvgQ313/HW0j2/LT+J/M9c8Pf97jYx3OOmxO/78V/89pXYBfAK+sesWcT8tL469jfwFG6cExiBzIOADA1pStvLfuPQD2p+8HytsIALe9kTpP60zHdzrS9q225h9ut3e7MfyT4UB51VB+ST5rD6/l822fc8t3t1T5uwpLClmTfPoN02fKntba8MCSB/jvivJnQuIS4wBYcXDFGR97w5ENZscDXynTZTXee60yL654kdZvtD71hvWMY5VuXWOpQBCUng7Nm596w2p4feTrjD23fGwiu4cHPezxMXam7SQ0MJSQgBCXdQF+AXRvarxac0SHERSXFRN3wLjYJGUlmb15ADYf30y7yHYoFC+tfInZf82m0zudWHZgmVOJwLHqoqi0iP/74f+czrktZZvT/JHsI2bVUH5xvtkG4e4FPZkFmSRnJQPw4soXGTRrkFPVV21w7KY7dv7YMzrWD7t/YGr81ErXv/3n2zyxtPwpcXu1nlKKZQeWVfu8OUU59P2gLzd+c2O1j1ET/J/zZ+yCM8tDTz2+9HEOZx/2qHNFfbL28Fq+2v6Vr5PhlqUCQWBGRrWrhU5HWGAYAApFg8AGLuufHvK02/22pmwlOjTa5WE1MHrpZBZmMqDVAO7ocwcAP+81nl1Izko2SwFgVAld0uES5lw1h6SsJG79/lYAvtj2Bcn5yeZ2765715yOS4xjxoYZTueseOFu9UYrUvNSAaON4GT+ScB9XXjbt9rS5k1j9NbjOUYV1fKDy93+7opDdp/MP8mR7CPmfEFJAeO+GsfO1J1u93eUV5xHxIsRzNsyz2l010W7F5FXnMdHf33EruzTHyNq/tb5vLb6NfOuOC4xju93fV/p9vaL2Lvr3mXYJ8NYvGfxaZ8Tyt95capgsuLgCqcSZU1ZtGsRq5NWAzh1WKjKvpP7nP79qqsu30FXZsORDfyZ7L5H3qjPRnHtl9dW+9hdp3fl5ZXeGajSeoGgqef98asrNDCUhwY+xNrbjQtpoF+g0/pnh7t26wSjjSCqQZT5TmRHecV5ZBZkEhkSaQaKE3kn8FN+ZBRksDPN+SLZuEFj+rfq77TsvfXvsejoIiKDIzm/2fnM2TQHrTXvrXvP7UX2m53fVFodkF+SbwaCii/oOZ5z3Ck4RIVGAZhVWY6WH1xO9KvR/Lj7R3NZs1eb0eqNVk7bfLHtC+5efDdgvBnOsU1kddJqM0DtSdtDTlEOkxdNZtlB54tnclYy//r+X9y18S63v6kqJ/NPkluca/6uEXNGcMWCKwD3vbPsPbnsXY4dA/XpsJe2qqpf1loz5OMh9JnZp1rnqMrYBWPNZzE8dc60c5z+/aqrPr4BsO8HfRk4y/XhUEfVaSsoKi1iV9ous0Re06wTCIqKCMzJqZUSgVKK10e+Tt+WfQHIeDSDvMfzeOyix5j595lu97EHi+jQaBSubRhZhVlkFGTQKKSRU4nhkg6XALBg6wJzqGyAxiGN3b53GaBzVGeGthtKSm4K8QfiuXvx3Tzw8wMu2y3Zu4Qle5e4PUZ+sUMgKMx0Chjf7vzWnC4oKTADhbunkVcdWgVgNkbnF+c79ZaC8gfzAvwCSM5KpvHLjbnvp/sAWLxnMYNnD+bJpU8CsC99n7FPcS7jvhrn8nsAynC9cKfnpzNvy7xKA5/9t7q703UMhPb9K/6HrW79uv3YFfPEkb3n2vHc4yzZu4TXV7/OG3+8UeVxVyet5qvkqqsp3AW4zIJM8w19NSU5K5lZG2e5LE8vqN7T73VBVf/e1Xmzob3UHNUgqtppqop1AkGqccdYG4GgotDAUBoENuC/l/yX2/vc7nab85qdB9gCga0xe0LPCeb6rMIsMgsziQyOdAoEU4dN5fLOlwPQp0UfM4g0adCERiGN3J7rjj53EBMWQ3pBOiPmjHC7zefXfA7Ak3FPul3vWDVkTx/ArI2z+L8fy9sa0vLSzLtad72Y7Bcb+wNq9moIKP/PdDjbqPIICQgxG7/t7SNzNs0Byi/Ue0/udZtewKzKCfMPc1k35ZcpjP9mfKWN2vbjH84+7HJHl5ZfXrVlb5eoeDdbWX339hPbKx2cEMpLBBW3cbzQHMs5Zk6P/mw0U36dwsO/PMyADwdUWr0yePZg3t33bpUXLHspy1Gjlxvx3xX/ZdWhVahnlUtg9DTgncw/aZaWxs4fy22LbnMZb+tUw6AcyDjA0eyjHp2vNji20zn+m1RUnQBn/xuzl65rmnUCQYrtj6wWqoY8cf351zvND2hlDAVRqkvp1LgTAE8Necpcn1WYZVQNBUcSERQBQERQBINaDzJfo5lTlGOWCho3aOz26d+Xzn+JW3vf6lR6qKjkqRKuO+86Av0C2Xh0IwDnRp3rtI1j1RCUV4F8sf0LALNhu/WbrdmVtsvcpri0mEd/e5Tnlj2HelaxcNdCoPwO2vFCbr+o2ruyaq05lHkIMKqf5m6ey9Ec40KQXWRc8Pad3Ffp71qVZJQ+CssKzV4w478Zz9LEpeZ/zt8TfweMu9TvdpYPIOhYInC8+JWUlTi1cdinKwYCx0EH7f5I+oPz3jvPDG4VbTy60fy9jiWCp5Y+Rad3OpGal8r6I+srfRHS2sNrXarHKnI3kq6dYxdiRxuObuCV1UYPNnuJzu5Uz5eczD/J4azDNH+tudmGZK82s7cl2VW8YK5OWs2v+34FjL+FDm93oOUbLatsVC4tK2XcV+Nc0jln0xx+3/87RWVFvLrqVRKOJRDzWgy7UneZA0uejpyiHGLfjzXnq7ohqc44X/a/q8pK+WfKOmMNnThhfPugRODOZ1d/xpwrjcbcw1mHiQmP4f0N79MqohVPXPwEP+z+gS5RXZg9djYP/vwgqXmp5Jfk0yikEUopdty9g9YNW6OUMp9l6NC4A7nFuRzPPU7jkMZO53vkwkcI9g9mgN8AlFJOgaBbdDd2pO7gX7H/YtYV5UV0+8W5eXhzVtyygmavle+z9vBapz/KXWm76NSkE7lFuVzY5kJeGPGC2eXU3uhcUlbCT3t/cnoz2/oj64Hy0oL9wg7GH39ySbI5DtOJvBPmBT+zMJObvr2Jc5qcA5QHor3plf8HtD+XUaJLOJ5znAC/AOZtmce8LfPMBv7PtnzGfwb/h27vdiOnKIfE+xNp07CN+Z/3cNZh84E9gCEfDTHTBMadW5vINi6B4Mm4J9l6YivTR08nKjSK/OJ8ftprdDPenbYbMC5OnRp3YnDbwSQcS6DPzPISnv1idyznmNm1+Jx3ziGzMJP5/5xf6W8+1R1zSm4KcQfi6N28t/lqVjt7r6+KNh3fZK6z32zEH4jnrh/v4opzrzC3+2DDBy4l4Dt+uIPFexabf1uO1UxHc47SI6Z86JaKeWhvq9h/337+SP7DXP7z3p8Z3Xm027QezTnKF9u+4Lud31HwZPlzORMXTgTgtg638WHih0SuiCSzMJOu73YFYNmkZS7PCFWlYmlmV9ouLm53MQnHEly2rU6JwF4681bVkHUCgb1qqI6UCPyUH37+fnRs3JGOjTsCsP2u7bSNbEtYUBgTY40/1Ft638LGoxuZvm46AJEhkQB0je5qHqt9o/Ysm7SM2OaxjP9mPBuPbqRJA+cnGW+JvYVuTbsRHx8PQEx4jLmuaVhTdqTucPpPCEaw+nX/r3x0xUcu6bffXYcHhZNTlMM/5v+D7k27k1uUy0VtLyImLMZlH6DSnjb2orTj3fbGoxu58vMrzfmU3BT8lb/TfvY7LzMQnNzLgFYDKh1LKcAvgJKyEq78/Epu6nmTudw+ntPO1J08E/+Meac8d/NcLu14KZryairHhl/HCxJUXiIAox1nZ+pOvrv+O9q9Vf6q1OKyYrTW5sXp+JTjvP7H6wDmee0cew/Zq43+SHJOgyN7aczut/2/OTXMJ2UlccWCK2gR3oIjDztX81QWCByX26ssvtj2BTtSd7AjtbwdaPIPkxncdjCtG7bmw40fcmvvW/l9/+9OpQbHkWGnr53OW2veMucru3Pu+E5Hp/ndabvNQPDD7h/4cfePFJcV8+ywZ81SRmFpecBxrC7bn2M8X1OxJFNZaaii0rJSCksLXUozyw8up2VESy6fd7nLPiPnjsRP+VGmyzj4wEHaRrY95Xmkaqim3HADyxcvhnPO8XVKKtWtaTfCglzrrx3r+iODI93uO6TdEBoGN6RVhNFbo3ED5xJBxdFL7SWCQL9A8+J1frPznba5sceNboOAo9HnlN+JbT+xnYOZB2kV0cqlCGsftruyQLA0cSnNXm3GBxs/MJc5BoHo0GiO5xyvtFviidwTFJQUkJSZxMVtL640veN7jKdJUBPWHl7LvT/d6/Jb/JQfH278EICOjTvyVNxT5tPcYHQHvXnhzZUe/2DmQdLy0jiRd8Lt+oRjCU4XOzAuxo53iT/v/dnt0Nk7U3e6Hbeqquqf7Se2c8/iexj31Thyi3L526d/460/y8+/4cgGwLkkZudYvdG5SWc23bmJm3s5//Y7friD9UfWO7XtOPp006c8/vvjPPzLwwz9eKjL3fDGoxvNYLdo9yKzlASnvnO297Bz/Jv4x/x/MGPDDGb9NYsLZ1/otq7e3qEAYFuW8axMxcb9Cd9OMPOmMr/v/52+H/Ql5rUYnop7ymndp5s/dRsE7OwlvN/2/+bULvXX0b+4csGVqGeV03s/vF0isE4gAMoaNAB//1NvWMc4jl/k7hkDR/ZAULFEUHHefsc+vud4esb0BMrr9T01+YLJfDj2Q9c0NGzlcj77sBsn8k5wYw/3D0fZL54V/9hfufQVHrnwEXKLc8ktznVZ36NZDwpLC5nwzQQ0mtjmsZWmuUtUF6bHTuflS537Y9/U8ybeGPkG7SLbcSLvBI1DGvP4RY9XehxH4UHhjD13LIF+gdz6/a1EvxrtNmD9eKNxJ/7mmjedli/es9jpTj/+QLzb92GPmjuKhbsW0r5Re6flm45vctnW7qe9P/Huunf5YtsXdHu3m8v65YfKn+1YcXAFv+z7hWUHlpFXnMf21PKn1VtGtKRnTE+iG7jWUff7oB9bUrbQLrKdy7pVSavMcbIqpjMyOJJ1R9ZV2qCdlpfGkewj5BXnue1y2TKiJS0jWnIk5wilZaUuF/1DmYecukW/uOJFAKeHuo4Xum9fAVyemC8oKWDJ3iXcu/hexnw2hks/vZSEYwnkFOWYbUsAQ9sNrfSYFb2/4X0GzRrEtzu+5ZVVrzBx4UTz5VZPLH2C2BmxrDq0ynxY0d1zSTXBOlVD9VjvFr2ZNnoaxaXFXNrx0iq3nRQ7iajQKPNCvHzScpYmLnVpOI4IjmDfffto3bA1hSWF3NXvLlpGnN57Gt7/x/tO8y3CW3A05ygxYTH4+/mzZPwS3v7zbX7a+5NTALu669VsP7HdrD898tARsouyOZBxgJFzRzr1wvnf5f/jzr53mo2EYJRcHO+Cu0R1YUvKFr7e8TUAnZp0qjTNgX6BtGjQgmsHXct/fvsPAKVPl5r50yWqC4kZiUY1W8/xpOSmMKD1ACYtnESXqC5O/+HtUh9JJTggmIgXIyguKqZFeAtGnjOSkrIS5m6ea253SYdLmHf1PHak7uD55c5Dj1z9RXmwn50wG4Bg/2AKSwtpEd6C2Oax5t3y4hsXM2beGKC8au79v7/P4azDRIZE8vAv7p9mT8pKYtQ5o5y6BDuW0IZ87FwnHuAXYFZhtIgwXuZk/7u6sM2FTqWAYP9gXvnbKy5ddlccMobYuO686/hi2xdO63q36O1UAqxow9ENdHu3G/7Kn/hJ8S7rg/yDjCqt7CM8v/x5t8Ou2y+qbRq2YcaGGbRq2IoXVrzgtM2UQVN47Y/XXPYN9Dc6S7y6+lXS89NdXj7VM6Yna25dw5rkNU697wa1HuS2lOauytLefub4729nbz8b+rERWCreANQkS5UI6rN7+t/Dg4MeNLuWVqZNZBvu6lf+wNTF7S7mmWHPuN22Y+OOBPkHEREcwUVtLzplGpaMX8K8q+dxVderXLaPCYth9a2rubrb1Qxtb/zhjjxnJB9d8RGxzWO58fwbmXPlHIa0G8Koc0bx1x1/UfZ0GQVPFNAiogVdorowrP0woPzJ64vbXsydfe8E4IIWF5jnqtiI99zw57iq61UE+wfTt2VfejTrYT7DYff2qLcBY3gOwGnwQMcgae8Ge3W3qwkJCOGxix9jRIcRHHrwEOc1Pc/czn5BDAsMIzggGICBrY0HiTZM3mD87phYpzQEBwRzQ48beG74c2ZgfHDgg07bvDXyLXPanseRIZG8NcpYfmvvWxl1zihzm6U3LyV+YjyT+0zm2eHP8tCgh9h9z24zbY9f9DhdorqYbSsTekzg2u6ePd1aUlbCoNZGtVjLcOMmwd57qeJd7+0X3M4V517hlEeO7h9wv8syewlicJvBfHLlJ9xw/g3mOoVi3ZF1ZBVmkV6QzosrX3R73JYRLYlLjGP+VqPB3PHf1V/5s+LQChqFNOK+AfdxKPMQExdOJNg/mP+OMMaFGtd6HHf0vcPtsTce3UifmX34fOvnbt9AOKTtEBoENjD/bu0cH+T8dty3hAeFs3HyRtbctoZ/xf6Lu/qe3gONpbqUYe2HsWFy1VVVZ8KrgUApNUoptUsptVcp9aib9Uop9Y5t/Wal1AXujiPqhpHnjOSGHjfwzbhvWHFL+WBqKVNS2H3vbto3as/X133tdPcfEx7DX3f8xeVdLuemXjexbNIysx1EKWVeRMG4wyt9upRnhz/Ln7f9yeLx5cMyODaS2auGBrQawJGHjtC9aXe+GfcNBU8WsO72dYQFhbHu9nWUPl1K1qNZ6Gc09w24j7Kny+jTsvzp23337WPj5I1Ov/Hfg/9Nq4hWLnXhUN7g2L9Vf1b/y7gbfunSl8z18/85n/iJ8ebds/05kKahTVky3vnBPHubyQMDH6D4qWLGdB5Dl6gu3D/wfnIfz2XW2FlM7jMZgCcufoIuUV1YectKpo2e5nQz0K9VPzPw2tnblPz9/HnhkhfYdc8uM6gMaz+Mz6/5nH337WNe/3kE+wdTmYigCCb2Mhqw7b/J3vOqQYBzFcWEnhMIDghmy/9tMQdddPw7GNR6EG9c9gZrbjXqw+/rfx+3X3A7Q9oNYeH1C7m5183M+2f5uz3sF9MLWlzA0HZDWbB1gbnOfm6NJiQgBI1md9pu7h9g5J3d8A5Gr7WYsBinG4ljU47x2MWPkTIlhTs73ck5Tc7h1b+9yup/rSYkIIRX//Yq9/S7x9z+5Utf5viU4zQPb06gXyBjOo/hkg6XMOXCKQAuN2eOv/vKrleS/Vg2vVv0BmDWFbN4YsgTTtt3i3ausps2ehrHpxxnzpVzzGWTek1yqW6tUVprr3wAf2Af0BEIAjYB3StsMwb4CVDAQODPUx23T58+urri4uKqve/Zoj7nweZjm/X+k/v1idwTeshHQ/TBjIPVOk5182DuprmaqejVh1Z7vM+qQ6v00eyjLssPZhzUb695W5eVlZnLSstKnbYpLSvVe9P2uj3u0v1L9aZjm9yuKy0r1ZMWTnJK57HsY3rhjoVO28XFxem0vDS9/+R+/UnCJ3ritxP1hG8m6LhEY3l+cb7ecnyLZip6/pb5Wmutt6ds1yH/L0TvSdujv9v5nZ7+53SdmJ7ocv5XV72qT+Se0C8sf0HvSt3ltD67MFuXlJa4TfuMdTP0LQtv0R//9bFmKnrm+pl60a5FmqlopqK/2vaVPpx1WDMV/dnmz5zWzUmYo7XW+rVVr+k1SWv0vM3zNFPRb695W2cVZOl2b7bTX2770iUP3Nl4ZKMePXe0/jP5T/PfpaC4QBeXFrvdfuvxrfrmb2/WTEUfyTqig58P1u3ebOd225LSEjPNl316md5/cr+++dub9QvLX9AvLH9BZ+RnmNtGvxKtmYpOyUlxe6zTAazXlV2vK1txph9gEPCzw/xjwGMVtnkfuMFhfhfQoqrjSiA4M5IH1c+DsrIyfSjjUM0mxoc8yYeysjK9YMsCXVBc4P0EOSguLdZzEuboopIiXVZWpp+Je0Z/te2rStO4/MByt8HFXRB2VJP/H8rKysx8yinM0XlFeZVuuz1lu84qyDrlMTcd26Q//uvjGklfVYFAaS+NMa6UugYYpbW+zTZ/EzBAa32PwzY/AC9prVfa5n8H/qO1Xl/hWJOByQAxMTF9FixYQHXk5OQQHh5+6g3PYpIHkgd2kg/WyoPhw4dv0Fr3dbfOm72G3LVqVow6nmyD1nomMBOgb9++etiwYdVKUHx8PNXd92wheSB5YCf5IHlg583G4mTA8Zn11kDFztWebCOEEMKLvBkI1gGdlVIdlFJBwPVAxcdKvwdutvUeGghkaq3rznCCQghhAV6rGtJalyil7gF+xuhBNFtrvU0pdadt/QxgMUbPob1AHlD1y2+FEELUOK8+Way1XoxxsXdcNsNhWgN3ezMNQgghqiZPFgshhMVJIBBCCIuTQCCEEBbntQfKvEUpdQI4eMoN3YsGXF/Eai2SB5IHdpIP1sqDdlprt2/mqneB4EwopdZX9mSdVUgeSB7YST5IHthJ1ZAQQlicBAIhhLA4qwWCmb5OQB0geSB5YCf5IHkAWKyNQAghhCurlQiEEEJUIIFACCEszjKB4FTvTz5bKKVmK6VSlFJbHZY1UUr9qpTaY/tu7LDuMVue7FJKjfRNqmuWUqqNUipOKbVDKbVNKXW/bbll8kEpFaKUWquU2mTLg2dtyy2TBwBKKX+l1F+2l2BZ7vd7rLJXl51NHzx4f/LZ8gGGABcAWx2WvQI8apt+FHjZNt3dlhfBQAdbHvn7+jfUQB60AC6wTUcAu22/1TL5gPHSp3DbdCDwJ8Z7wS2TB7bf9RAwD/jBNm+p3+/pxyolgv7AXq31fq11EbAAuMLHafIKrfVy4GSFxVcAn9imPwGudFi+QGtdqLVOxBgOvH9tpNObtNZHtdYbbdPZwA6gFRbKB23Isc0G2j4aC+WBUqo1cDnwocNiy/z+02GVQNAKSHKYT7Yts4oYbXvhj+27mW35WZ8vSqn2QG+MO2JL5YOtWiQBSAF+1VpbLQ/eAv4NlDkss9Lv95hVAoFH70a2oLM6X5RS4cDXwANa66yqNnWzrN7ng9a6VGsdi/EK2P5KqfOr2PysygOl1N+BFK31Bk93cbOs3v7+02WVQGD1dyMfV0q1ALB9p9iWn7X5opQKxAgCn2mtv7Ettlw+AGitM4B4YBTWyYPBwFil1AGMquARSqm5WOf3nxarBAJP3p98NvsemGibngh857D8eqVUsFKqA9AZWOuD9NUopZQCZgE7tNZvOKyyTD4opZoqpRrZphsAlwI7sUgeaK0f01q31lq3x/j/vlRrPQGL/P7T5uvW6tr6YLwbeTdGb4AnfJ0eL/7O+cBRoBjjLudWIAr4Hdhj+27isP0TtjzZBYz2dfprKA8uwijWbwYSbJ8xVsoHoCfwly0PtgJP25ZbJg8cftcwynsNWe73e/KRISaEEMLirFI1JIQQohISCIQQwuIkEAghhMVJIBBCCIuTQCCEEBYngUAIN5RS7R1HcPVg+0lKqZYebDP9zFMnRM2SQCBEzZgEVBkIhKirJBAIUbkApdQnSqnNSqmvlFKhSqmnlVLrlFJblVIzleEaoC/wmVIqQSnVQCnVTym12vY+gLVKqQjbMVsqpZbYxsN/xYe/TQiTBAIhKncuMFNr3RPIAu4Cpmut+2mtzwcaAH/XWn8FrAfGa2OQt1Lgc+B+rXUvjOEd8m3HjAXGAT2AcUopx/FthPAJCQRCVC5Ja73KNj0XY+iK4UqpP5VSW4ARwHlu9jsXOKq1Xgegtc7SWpfY1v2utc7UWhcA24F23v0JQpxagK8TIEQdVnH8FQ28B/TVWicppaYCIW72U272tSt0mC5F/g+KOkBKBEJUrq1SapBt+gZgpW061faug2scts3GeC0mGKN8tlRK9QNQSkUopeSCL+os+eMUonI7gIlKqfcxRqv8H9AY2AIcwBje3O5jYIZSKh8YhNEOMM02BHQ+RjuBEHWSjD4qhBAWJ1VDQghhcRIIhBDC4iQQCCGExUkgEEIIi5NAIIQQFieBQAghLE4CgRBCWNz/BwhhdD7VuNP3AAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "history.loss_plot('batch')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "5285f730",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "10000/10000 [==============================] - 6s 554us/step\n",
      "[0.04901809926041169, 0.9842000007629395]\n"
     ]
    }
   ],
   "source": [
    "#评估模型\n",
    "score = model.evaluate(X_test, y_test)\n",
    "print(score)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "51b200fd",
   "metadata": {},
   "outputs": [],
   "source": [
    "y_pred=model.predict_classes(X_test)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "d50179c9",
   "metadata": {},
   "outputs": [],
   "source": [
    "from sklearn.metrics import confusion_matrix\n",
    "# 加载相关库函数\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "def plot_confusion_matrix(y, y_pred, cmap=plt.cm.Blues, title='混淆矩阵'):\n",
    "    \"\"\"\n",
    "    绘制混淆矩阵\n",
    "    :param y: 真实值\n",
    "    :param y_pred: 预测值\n",
    "    :param cmap: 热力图的颜色\n",
    "    :param title: 图像标题\n",
    "    :return:\n",
    "    \"\"\"\n",
    "    cm = confusion_matrix(y, y_pred)\n",
    "    classes = list(set(y))\n",
    "    classes.sort()\n",
    "    plt.imshow(cm, cmap)\n",
    "    indices = range(len(cm))\n",
    "    plt.xticks(indices, classes)\n",
    "    plt.yticks(indices, classes)\n",
    "    # 热度表\n",
    "    plt.colorbar()\n",
    "    # 坐标轴含义\n",
    "    plt.xlabel('guess')\n",
    "    plt.ylabel('fact')\n",
    "    plt.title(title)\n",
    "    # 显示数据的值（Numbers）\n",
    "    for first_index in range(len(cm)):\n",
    "        for second_index in range(len(cm[first_index])):\n",
    "            plt.text(first_index, second_index, cm[first_index][second_index])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "6abb724f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAjgAAAHvCAYAAABZpqX8AAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjMuNCwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy8QVMy6AAAACXBIWXMAAAsTAAALEwEAmpwYAABYOklEQVR4nO3de3RU9b3//+ee3BMCyYTcgCCEAFaiRRJQLgJeoFilXKpcFBRvtP4Usfbb9T09Hi3VnvP9rvM9tSja1iyrgJdyKXJTwaAFImAIabAWRSGBcDEJJCRyCROSzMzvj2gUA7ngTPbsPa+Ha9bqbHZmv/LpzuSd9+ez9xher9eLiIiIiI04zA4gIiIi4msqcERERMR2VOCIiIiI7ajAEREREdtRgSMiIiK2owJHREREbEcFjoiPeDyei/6by+Vi/fr1zc9PnDjRYp8NGzbQ2l0bqqqq+P3vf8/Zs2d5/fXXW83yxRdfALB7927Kyso4evQou3btYu3atbz44ottfSvNTp06xYcffthiu8vlYvny5e16jQuNS3V1NRs3bmx3DhGRjlKBI+IDXq+Xp59+mqNHj7Jo0SK2b9/OwoULWbJkCQBRUVF88sknzftv3ryZ999//7zX2LlzJ8XFxbzwwgs0Nja2OMYzzzzDHXfcQVRUFCEhIRQWFvK3v/2NQ4cOtdj3X//6Fy+++CLdunXjv/7rv9i1axcej4fevXsTHh7eYv/ly5dz+PDh5q9ds2YNu3bt4r//+79JS0trsf+hQ4eorKwEmgqYN998kz179vDWW2/xhz/8gfLy8uZ9X3/9dXJzc/njH//ISy+9BMC+ffsu+D2KiPiKChwRHzAMgwcffJCamhoqKyuJjIxk3LhxJCQk8OKLL3LmzBm6dOnSvP9tt93Gxx9/fN5rJCQk0L9/f6688srm4uFrb731FqWlpSxbtowXX3yRzZs3c+rUKTIzM+ndu3eLPBMmTOCmm27C4XDQ2NhIWVkZhYWFHDly5IIFy9SpU9mwYQMAV155JefOnWPVqlVER0dTW1vbYv/8/HxmzJjRXKRs3LiRo0ePkpGRwc9//nNSU1Ob9/V6vTidTurr68nOzmbNmjX8/ve/5/PPP+f555/nj3/8I+fOnevAaIuItC3U7AAiduDxeEhKSiI8PJykpCQ++eQTMjMzAaivrycmJqbF1/ziF78AmgqAU6dOUVNTwwcffMAXX3xBSkpKc5Hw6aefctlllzFixAj69u3LwYMHqampoaysjC5dumAYxgUz9evXjwMHDvCjH/2IKVOmALBixQoGDBjQYt+wsDDmzp3Lvn37+Pjjj0lJSeHGG29k9OjRrFq1ioEDB563/86dO8nLy6NPnz5MmjSJq666igkTJgBcsIgaOHAg//znPxk0aBBdu3Zl//79XHnllYwfP74jwywi0m4qcER84JVXXmHEiBGUlJQA4Ha7mwsPwzCa/3ddXR0fffQR7777LklJSQwfPpytW7fSo0cP6urqcLvdjBo1ipSUlObXvuKKKwDYunUrISEhdOnShYiICBobG5vX2nzbypUriY2NpWfPnpw5c4b169ezb98+du/ezfXXX09UVFSL/a+//npKS0upra2lrq6OP//5zwwfPpySkhLq6+t55plneOihh4iIiODAgQPMmDGD/Px8/vf//t8APPfcczzzzDPU19dTVFTEihUrml/f4/FgGAbh4eE4HA7279/PjTfeyFtvvaUCR0T8RgWOiA/cddddFBQUcPr0aQAaGhrOm5L6tpqaGoYOHUrXrl0ZPHgwmZmZ7N+/n9jYWLp06XLBKSSA0tJSwsPDqauro3v37jQ2NlJZWcnOnTu55pprmve7/fbbefnll/F6vezfv5/nn3+e6Oho/uM//oOf/exn7Nq1i549exIa2vTjf/PNN7Ny5UrS0tLIysqitLSUBx98kEGDBlFeXs7Ro0ebuzMA586d44orruCvf/0rf/rTn0hJSeHqq6/m4Ycfxu12s3DhwvNyezweXC4XV155JTt27OCVV15h9OjR7Nmzh2effZbu3btz5513fp/hFxFpQQWOiA+EhYXhdDrxeDwcP36cc+fONXdt6uvrm/eLjIzk5ptvZuPGjcTGxvLxxx+zfv16HA4HM2fOJC8vj+zs7Ase45577mHQoEH8/e9/5/LLL+eyyy7jyJEj7Nmzh9DQULKyspr3vffee1m2bBkA0dHRnDp1ipSUFMrKyli1ahWXX345sbGxAHTp0oW7776b9evXU1ZWxksvvcTEiRP5/PPPSU5O5uDBg+cVOB999BFut5uYmBjmzJlDVFQUzz//PLW1teTn59OtW7fzcnu9XkpKSli6dCk/+9nPePnll4mOjsbj8fDwww/75v8AEZHvUIEj4iObNm3iwQcfZNu2bURERBAWFkZMTAwZGRkt9j19+jT9+/enX79+REVFUVpaSp8+fVi9enWLfY8cOcLatWuJiIhg9erVvP766wwYMID77ruPQYMGMWzYMLp27Xre1xw7dozo6GjOnTvHc889R0JCAuXl5Wzfvp3/+3//b4uFw3l5eURFRfGvf/2L6dOnYxgGdXV1REdHExISct6+M2fOxOPxsGXLFo4ePUr//v0pKCjA7XbTt29fbrzxxvP293g8DB48mOjoaI4ePcr27dupq6tj69at1NbWEhISQnx8PHPmzGlxLBGRS6UCR+R7crlcfPLJJ5w8eZKwsDDmz59PTk4OO3bs4Pbbb2+eCvq206dPExoaSn5+Pl27dmX37t107dqVtLQ09u3bR1hYGPHx8cTFxZGWlsbDDz/MW2+9hdvt5ve//z3p6el8+OGHHDlyhH79+p332l988QW5ublMnDiRw4cPN3dJnn/+eW6//XYACgsLGTNmDG63m127dlFQUMCPf/xjxo8fz/PPP4/L5SI0NJTq6uoLfs/vv/8+//Zv/8aKFSuai6yLdWM8Hg8Oh4OrrrqKvn37Mm7cOMLCwggNDeWRRx75PkMvInJRukxc5HuKiopi//79JCYmApCTk8OcOXMYOXIk/+///T9KSkooLy/H5XI1f011dTVLly7lk08+4b333uPOO+/kmmuuYerUqbz88sts3769eQ3Prl27WL58OQMGDODnP/859fX1hIWFcc8993DkyBHmzZvH22+/3XzJ9vvvv8+qVaua1+m8++67vPnmm5w5c6b5+CdOnODs2bOEhIRQX1/PgQMHmq/68nq9/K//9b+49957mThxYourtNxuNxUVFWRkZPDv//7vDB48mHPnznHo0CEKCwtZt24dS5cubc7z9SJjgNjYWMLCwpqPIyLiL+rgiPjAzJkzaWhoYMWKFUyaNIm4uDji4uL45S9/SXh4OE899RRJSUnN+/fo0YNbb721xdSSw+HgV7/6Fffeey+TJ08mIiKC9PR0hg4d2rxPUlIS0dHRAMyePZvZs2ef9xp33XUXPXr0AOCzzz6jZ8+e3H777bzyyitcd911dOnShZiYGEaOHEl0dDSjR4/m6quvbv76r4uRLl268Ic//IG+ffs2/5vX62Xt2rVMnz69eVtcXBzl5eXs2LGDlJQUMjIySElJae5cnT179oKXsn+9IFtExB8Mr/6MEvGJc+fO4Xa7m4uPb/N6vdTV1bW4RLs1Xq/3ove4aa+GhobmjglAY2PjBafMWjvut7+moaEBj8dDREREuzPU19df8O7Jp06dalHgiYj4igocERERsR2twRERERHbUYEjIiIithMQi4yNiFgcMQlmx2iXwX27mx1BREQC0KFDpVRVVZkdo0NCul6Gt9HV9o4dMH70EDZu3OjT17wUAVHgOGISiLzxN2bHaJftf73H7AgiIhKARl174buQBzJvo4uIgdN8+ppVVfk+fb1LFRAFjoiIiJjBAMOeq1Xs+V2JiIhIUFMHR0REJFgZwPe831agUgdHREREbEcdHBERkWBm0zU4KnBERESCma+nqALk8xHsWbaJiIhIUFMHR0REJGj54TJxdXBERERE/EMdHBERkWBm08vEVeCIiIgEKwPbXkVlz+9KREREgpqtOzje+loaD+ZhdEkCjxtHbCrnCv+Co0sK3oazhKaPIbRnFo1HdgIGnqrPCb96tqmZf/fUAuLi4klISGDmnbNMzdIat9vNa0uXEBcfz6ef7OHXjz9hdqQ2WWVswVpZwTp5D5WW8uDP7iehe3cAXvhTDl27djU5VetWrljO7dOmmx2jTRpbqzJsO0Vl6w5O44GthPQaSmjPLLxnjuGpPU7k2F8Tce3PCb1sBCE9rsZzthpvg4vQtGE44tPxes1b/r27qIjIqCgefmQ+eVu3UF9fb1qWtry3KZducXFMmjyF6OgYPtmzx+xIrbLS2FopK1gv7+NPLODV15fx6uvLAv4X8Ntvree1pYvNjtFuGlsJJLYucLxnqzCi4gAwIrthhERghEbgdTeA14NhOHCXFeGIuwyA0D4jMUysZHPf3cDw4SMBSM/IYFdBgWlZ2tKrVxqhod80ACMjI01M0zYrja2VsoL18lrJLbdOJCkp2ewYtqSx/RbD4dtHgLD1FJXRJQnPl4dxxPfFfaKEkKRwANxHdxGSPAgAb20VHk8jnqp9eM+eIOyHM00rcsrLykhMTATAGe+koqLclBztMSgzk0GZmQAcPHiA9H79TE7UOiuNrZWygvXyvv9eLoW7CqiuPsFvn/5Ps+PYisbWojRF1XELFizg2Wef5bXXXvPnYS4qNP16PJWf4z60DUdMd4ywaAA8NaUYkd0A8DbW4YhNJWzAj3B064Wnar8pWb/L6/Wa2k1qr5UrljP/0ccskfVrVhlbsFZWCPy8iUlJzLn3fub/4jFCQ0M5VFpqdiTb0NhKoPFbgVNUVERUVBTz589nyxZz5uWN0AjCBt5MaJ/r8DbUYcQm43U34D136pt9wmMxopxN/zs6AW/dl52e82upPXpQVVUFQE1NNSkpqaZlaY9dBQX06pVG3/R0s6O0yUpja6WsYK289fX1zWtDevbsxbFjx0xOZB8aW6sybDtF5bckGzZsYOTIpnn5jIwMCkyYl/ecOY674l9NTxrrcHRJxnu6HBxhzfuEJA7AU1MKgLfuJI7YlE7P+bVx4yeQn78DgJLiYoYOG2ZalrbU1tZSUlLM8BEjcLlcbN+2zexIrbLS2FopK1gr76tLFrPtgzwAysvL6NO3r8mJ7ENjK4HGbwVO2bfn5Z1OystNmJd3hOL+ooiG/ZsI6dNUbGE4MMJjvtkleRBeVzWNRwubLiWP6935Ob8yJCuLOpeLRc8uZPSYsYSFhbX9RSZZ8srLrFu7mtl3zmDcDWNwOp1mR2qVlcbWSlnBWnmnzZjJ8WPHWP3mKpKSkklKSjI7UqvWr1vL1q2beW9TrtlR2qSxtSiDpjU4vnwECMPrp+uiH3roIR555BEGDhzIiy++SEJCArfddlvzv+fk5JCTkwNA0d5Son78//wRw+dO/PUesyOIiEgAGnVtNoWFhWbH6BBHbA8iBt/v09cc5HonIMbBbx2cHt+al6+uriY19fx5+blz51JYWEhhYSFGRBd/xRAREZHWaA1Ox0yYMIEdO5rm5YuLixkWwPPyIiIiwUmLjDssKysLl8vFwoULGTs2sOflRURExF78eqO/J5980p8vLyIiIt+XI3AWBvtS4PSSRERERHzE1h/VICIiIq0wCKh1M76kAkdERCSYBdC9a3zJnmWbiIiIBDV1cERERIKWYdspKnt+VyIiIhLU1MEREREJZjZdg6MCR0REJJhpikpERETEGtTBERERCVaGYdspKnVwRERExHbUwREREQlmNl2DowJHREQkmGmKSkRERMQa1MEREREJWrqTsYiIiIhlBEQHZ3Df7mz/6z1mx2iX+KEPmx2hQ2p2PW92BBERCWRagyMiIiJiDQHRwRERERETGNh2DY4KHBERkaClRcYiIiIilqEOjoiISDAzYZHx8uXLmT59OgALFiwgPj6ehIQEZs2a1aFtrVEHR0RERDrN+vXrWbx4MQBFRUVERUUxf/58tmzZQn19fbu3tUUFjoiISDAzHL59tGHixIkkJycDsGHDBkaOHAlARkYGBQUF7d7WFk1RiYiIBDMT74NTVlZGYmIiAE6nk/Ly8nZva4sKHBEREfGZyspKsrOzm5/PnTuXuXPntvl1Xq8X4zvFVnu3XYgKHBERkWBl+P4y8cTERAoLC9u1b48ePaiqqmLgwIFUV1eTmZnZ7m1t0RocERERMcWECRPYsWMHAMXFxQwbNqzd29qiAkdERCSYGYZvH21Yu3YtmzdvJjc3l6ysLFwuFwsXLmTs2LGEhYW1e1ub35bX6/X6Yny+j6ysbLbvbF87y2z6sE0REbmQUddmt3tqJlA44vsQecOTPn3NKw79MSDGQR0cERERsZ2gWGS8csVybp823ewY7eKu2U9IfP8LPvc21uE+eQDDCMHr9RCa8AM8507ReGQzhEYCEJZ2PUZIuN9znjx5kuV/fYOU1FSqKiu59/4H/H7M7+t3Ty0gLq7pLpgz72z7LphmslJWsM7PWE1NDWveXEVERARut5vZd88xO1Kr3G43ry1dQlx8PJ9+sodfP/4EK5YvwzAMtn+Qx8JFL5gdsZnVxhYI2LHsTAa064okK7J9B+ftt9bz2tLFZsdoF/fJg7irP7voc09tGUZIOCHOgXjOlDVvD00ZSnifHxHe50edUtwAvP7qUqbPvIOfTJpMvNPJJ3v2dMpxL9XuoiIio6J4+JH55G1t310wzWKlrGCtn7FtH+TRLS6OO2bNJm/rFrPjtOm9Tbl0i4tj0uQpREfHsH3bNk6dPMnt06aTPXQYAbDCoJnVxvbo0aMBO5biG7YvcG65dSJJSclmx2iXkG59ITS6lefpOLr1A8BwmPt/XWxsLO9u3ABA9YkTdO3WzdQ8bcl9dwPDhzfdBTM9I4Nd7bgLplmslBWs9TM28SeTmDL1pwCEh3fOHwPfR69eaYSGftNo/+dHuxl89RAAZt11d0D95W21sV23ZnXAjmWnMvzwCBBBMUVlK54GGsrzmwsdAPfpI3jOHsfrriMs9dpOiXHHrNnMmPZT3t+Uy6jrRpOWltYpx71U5d++C2a8k4qKtu+CaRYrZbWiM2fO8OR//DuTv/plHMgGZWYy6Kv7fRw8eICIiAjqG+rZvu0DDh8+xP88szCgfjFbaWwPHSoN6LGU78+vbYDly5f78+WDkhESTliv0XhOleJtOIsRFkVowhWEJg0GDDznTnVKjs/27mXy5Kn86OYfs3TJKwE/jfJt7b0LZiCwUlariI2N5Q/PLuKdt9/i+PHjZsdpl5UrljP/0cc4c/o0AwdezvxfPMaVV17F9u3bzI52HiuNbaCPZecxMAzfPgKF3wqcb39aqPiGt7EOr7upkDAinXhqy8DjAUdTO9gI7wKNZzsly/JlbzBtxkym/vQ2pky9jfc25XbKcS9V6ld3wQSoqakmJSXV5EQXZ6WsVlNTU8OpU01/BAwalMkHeVtNTtS2XQUF9OqVRt/0dBK6d6dXr6ZuaVrv3pSXlbXx1Z3HamMbyGPZ2VTgdNC3Py1UfMNd8zmeU4eanjSexQjvirv6s6ZCB6ChFiO8a6dk6RYX13wi9+jZk6ioqE457qUaN34C+flNd8EsKS5maDvugmkWK2W1mtdfXcrGDe8AcOxYBX37ppucqHW1tbWUlBQzfMQIXC4X0dHRFP2j6f4ixyoqGDBgoMkJv2G1sR113eiAHUvxDdNWqubk5JCdnU12djaVVZV+O876dWvZunVzwHcYANwnD+A5cxT3qcMXfB4S1x9vowv3l8UQEoEjOomQ+P54G8/i/rIEQqMxwqJbO4TPzLnnPl5buoR1a9dw5PBhxl5/Q6cc91INycqizuVi0bMLGT2mfXfBNIuVsoK1fsZunz6DqspKVv1tJXFx8QzJyjI7UquWvPIy69auZvadMxh3wxgmTZ7K0aNHWP3mKhoaGvjh4MFmR2xmtbG9adz4gB3LzmbXDo5f72Q8Z86cdk1T6U7G/qM7GYuIdA4r3sk4xNmXmPG/9elrDih+LiDGQVdRiYiIBLFA6rr4kgocERGRYBVg967xJb+twfn2p4WKiIiIdCa/dXAmTZrEpEmT/PXyIiIi8j0ZBNbCYF+y/Uc1iIiISPDRGhwREZEgZtcOjgocERGRIGbXAkdTVCIiImI76uCIiIgEMXVwRERERCxCHRwREZFgpRv9iYiIiFiHOjgiIiJBzK5rcFTgiIiIBCndyVhERETEQtTBERERCWLq4IiIiIhYhDo4IiIiwcyeDRwVOCIiIkHL0BSViIiIiGWog9NB1QWLzI7QIfGTrJO3Zu08syOIiAQddXBERERELEIdHBERkSBm1w6OChwREZEgpTsZi4iIiFiIOjgiIiLBzJ4NHHVwRERExH7UwREREQlWutGfiIiIiHWogyMiIhLE7NrBUYEjIiISxOxa4GiKSkRERGxHHRwREZFgZs8Gjjo4IiIiYj+27+D87qkFxMXFk5CQwMw7Z5kdp02/e2oB8fHxOJ3WyOsuKyKkxxAAGva9gxEWjREeQ0jPoRfcp+n5PwADT3UxYZnTOi2rlc4FK2UFa+W1UlaAlSuWc/u06QCsWL4MwzDY/kEeCxe9YHKy89XU1LDmzVVERETgdruZffccsyO1yWrngr9oDY4F7S4qIjIqiocfmU/e1i3U19ebHalVu3cXERUVxUPz5pOXF/h53cf+hfuLnQB4Th7BCAkjtO9YPCf24/U0ttgHwOuqwdtYR0iPITjiLsPr9XZKViudC1bKCtbKa6WsAG+/tZ7Xli4G4OjRo5w6eZLbp00ne+iwTvvZaa9tH+TRLS6OO2bNJm/rFrPjtMlq54K/GIbh80egsHWBk/vuBoYPHwlAekYGuwoKTE7UutyNGxg+oilvv34ZFO4K7LwhyVdCeCwAnspPccSnA2DEJOL98lCLfQDcxz7G0bVX07/1uqbTfhisdC5YKStYK6+VsgLccutEkpKSAVi3ZjWDr27qhM666+6A+kUCMPEnk5gy9acAhIeHm5ymbVY7F6TjbD1FVV5WRmJiIgDOeCcVFeUmJ2pdeXkZ3bs35Y13OqkoD+y83+atOwnhXZqehEXjPXfqwvu5qvF43HiqS/DW1RD6g6md8kZtpXPBSlnBWnmtlPW7Dh0qpb6hnu3bPuDw4UP8zzMLA67IOXPmDE/+x78z+atCJ5BZ+VzwtUA7j3zFbx0ct9vNyy+/zOrVq3n66af9dZh283q9lvo/0Wp5263xHEaXZELTb8CI7YG35kCnR7DS2FopK1grr5WyApw5fZqBAy9n/i8e48orr2L79m1mR2ohNjaWPzy7iHfefovjx4+bHafdrHYuSPv4rcDJzc0lLi6OKVOmEBMTw549e/x1qItK7dGDqqoqAGpqqklJSe30DB2RmtqDEye+yltdTUpqYOf9NiOyG9TXNj2pr8WI6HbhHcNjMCLjvvqa+KbOTyew0rlgpaxgrbxWyvpdCd2706tXGgBpvXtTXlZmcqLz1dTUcOpUU+d20KBMPsjbanKi1ln5XPA1rcHpoLS0NEJDv5kBi4yM9NehLmrc+Ank5+8AoKS4mKHDhnV6ho4Y96MJ5H/4Vd6SYrKHBnbeb3Mk/gDPlwcB8J6twoi77ML7OfvhOXm46Un9KYwuSZ2Sz0rngpWygrXyWinrd426bjRF/ygE4FhFBQMGDDQ50flef3UpGze8A8CxYxX07ZtucqLWWflc8DnDx48A4bcCJzMzk5/85CcAHDhwgH79+vnrUBc1JCuLOpeLRc8uZPSYsYSFhXV6ho4YMiQLl8vF888tZPTowM/rPvYxnhP7cVfuxdGtN7jraTy4GUdCBoYjpMU+AI7ul0Pdl7jLP8LrcTcvOPY3K50LVsoK1sprpawA69etZevWzby3KZebxo3n6NEjrH5zFQ0NDfxw8GCz453n9ukzqKqsZNXfVhIXF8+QrCyzI7XKaueCdJzh9fO1hsuXL2fo0KGkp59fzefk5JCTkwPA8cpK9pUc8mcMnwm0SzPb4pz8vNkR2q1m7TyzI4iIXLJR12ZTWFhodowOiUjuT887n/XpazrzngyIcfDrZeIFBQWkpaW1KG4A5s6dS2FhIYWFhSR+deWQiIiIiC/4rcCpra2luLiYESNG4HK52LYt8Fb8i4iIBDVDi4w77OtLxGfMmMGYMWNwOp3+OpSIiIjIefx2o7958+Yxb57WVIiIiAQqAwigpotP2fpOxiIiItKawJpW8iVbfxaViIiIBCd1cERERIKYTRs46uCIiIiI/aiDIyIiEsTsugZHBY6IiEiwMjRFJSIiImIZ6uCIiIgEKQNwOOzZwlEHR0RERGxHHRwREZEgZtc1OCpwREREgpiuohIRERH5nk6ePMkbb7xBamoqlZWVPPDAAyxYsID4+HgSEhKYNWsWwAW3dYTW4IiIiASrry4T9+WjLUuXLuWOO+5g8uTJOJ1O8vLyiIqKYv78+WzZsoX6+nqKiopabOsoFTgiIiLSaWJjY9mwYQMAJ06cYPPmzYwcORKAjIwMCgoK2LBhQ4ttHaUCR0REJEgZNK3B8eWjsrKS7Ozs5kdOTs55x5w9ezbLli3j3nvvJTw8nOPHj5OYmAiA0+mkvLycsrKyFts6SmtwRERExGcSExMpLCy86L/v3buXqVOnEh0dzaJFi7jiiiua/83r9bZY9Hyhbe2hAqeDrLbavGbtPLMjtFv8lD+ZHaFDalY/aHYEEZHvyej032tvvPEGv/3tbwkLC6O8vJzq6mqqqqoYOHAg1dXVZGZm0qNHjxbbOkpTVCIiIkGssxcZx8XFNRdVPXv2pEuXLuzYsQOA4uJihg0bxoQJE1ps6ygVOCIiItJp7rvvPpYsWcKaNWs4fPgwjz32GC6Xi4ULFzJ27FjCwsLIyspqsa2jNEUlIiISxDp7iiohIYH77rvvvG1PPvlki/0utK0j1MERERER21EHR0REJFi1c92MFanAERERCVJf3wfHjjRFJSIiIrajDo6IiEgQs2kDRx0cERERsR91cERERIKYr9fgeH36apdOBY6IiEgQ8/UUVaAUOJqiEhEREdtRB0dERCRYGbpMXERERMQybN/B+d1TC4iLiychIYGZd84yO06brJZ35Yrl3D5tutkx2sVdVkhIj2wAGj5/CyMsGiM8hpBe11xwm7fBhfuLXRiRXfGeO0PoZaM6LauVzoNDpaU8+LP7SejeHYAX/pRD165dTU51cVYaW7fbzWtLlxAXH8+nn+zh148/YXakNlnpPcFK54K/NN3oz+wU/mHrDs7uoiIio6J4+JH55G3dQn19vdmRWmW1vG+/tZ7Xli42O0a7uI99jPtIPgCek4cxQsIJTb8Bz4n9eD2NF9zmPppPSM+hhKQMxgiPwXOqrFOyWu08AHj8iQW8+voyXn19WUAXN1Yb2/c25dItLo5Jk6cQHR3DJ3v2mB2pVVZ6T7DauSAdZ+sCJ/fdDQwfPhKA9IwMdhUUmJyodVbLe8utE0lKSjY7RruEJF8FEU2/eD3HP8HhTAfAiEnE+2XpBbcRGonn+CcAeOtrMcIiOyWr1c4DK7Ha2PbqlUZo6DeN9sjIzjkHL5WV3hOsdi74j4Fh+PYRKGw9RVVeVkZiYiIAzngnFRXlJidqndXyWpW37iSExzY9CYvBW3fqgttCel1DQ2EOnqq9GM4MjChnp+Sz4nnw/nu5FO4qoLr6BL99+j/NjnNRVhvbQZmZDMrMBODgwQOk9+tnciL7sNq54E8BVJP4lN8KnJqaGlatWkVERARut5s5c+b461Dt4vV6A6qybIvV8lrXhe7Y0LTNe7qCkNTB4AinsXQLIT2HYjg6928CK5wHiUlJzLn3fnr37s3Tv/0Nh0pLuaxPH7NjtckKY/u1lSuWM//RxyyT12qsdC5I+/ltiiovL4+4uDhmz57Nli1b/HWYVqX26EFVVRUANTXVpKSkmpKjvayW16qMyG5Qf6bpSUMtRmS3C25zl+3C0WMoIT2GEJJ6NZ7KvZ2Sz2rnQX19ffO6m549e3Hs2DGTE12c1cYWYFdBAb16pdE3Pd3sKLZixXPBX+w6ReW3AmfSpEn89Kc/BSA8PNxfh2nVuPETyM/fAUBJcTFDhw0zJUd7WS2vVTkSB+GpOQCAt7YSI67PBbcZoVHNX2NExkFIWKfks9p58OqSxWz7IA+A8vIy+vTta3Kii7Pa2NbW1lJSUszwESNwuVxs37bN7Ei2YbVzQTrOr4uMz5w5w7x585oLnW/LyckhOzub7OxsKqsq/XL8IVlZ1LlcLHp2IaPHjCUsrHN+QV0qq+Vdv24tW7du5r1NuWZHaZO74p94TuzDXfkpjrje4G6g8cDfcSQMwHCEXHBbSO8RuI/m4674CK+rBkfCwE7JarXzYNqMmRw/dozVb64iKSmZpKQksyNdlNXGdskrL7Nu7Wpm3zmDcTeMwensnHVgl8pK7wlWOxf8xmhag+PLR6AwvF6v3z82Yt68eTzxxBMXfePLyspm+85Cf8eQABc/5U9mR+iQmtUPmh1BRALIqGuzKSy01u+y2LTLGfzoSz59TddfHw2IcfBbB6empoZTp04BkJmZydatW/11KBEREZHz+K3AWbp0Ke+88w4AFRUVpGuBnIiISMDRIuMOmjFjBpWVlaxcuZL4+HiysrL8dSgRERGR8/jtph7JycnMmzfPXy8vIiIiPhBATRefsvWdjEVERKR1gTSt5Eu2/iwqERERCU7q4IiIiASrALt3jS+pgyMiIiK2ow6OiIhIkDIIrEu7fUkFjoiISBCzaX2jKSoRERGxH3VwREREgpjDpi0cdXBERETEdtTBERERCWI2beCogyMiIiL2ow6OiIhIkDIM+35UgwocERGRIOawZ32jKSoRERGxH3VwREREgphdp6jUwRERERHbUQdHAkbN6gfNjtAhzhkvmx2h3aqX3Wt2BBEJUDZt4KjAERERCVYGTR+4aUeaohIRERHbUQdHREQkiOkycRERERGLUAdHREQkWBmGbS8TV4EjIiISxGxa32iKSkREROxHHRwREZEgZQAOm7Zw1MERERER21EHR0REJIjZtIGjDo6IiIjYjzo4IiIiQUyXiYuIiIitGIZ9p6hsX+D87qkFxMXFk5CQwMw7Z5kdp01Wy7v3009Zs3oVv378CbOjtMkqY+utr6Xx4Ac4uiTh9TQSkpKJ+/BOjMhueOtPE9p3DF6vB/eh7RhhMXhOfUHYDyaamtkqY/u1lSuWc/u06YA1slvh58ztdvPa0iXExcfz6Sd7AjorwMmTJ1n+1zdISU2lqrKSe+9/wOxI4mO2XoOzu6iIyKgoHn5kPnlbt1BfX292pFZZLS/A+nVrcLvdZsdok5XGtvHgVkJ6ZRPScwje2uM0Fr9PSO9rCOk5BCM8Bs/Jo3iOfYIRFk1IzyEQEo7n5FHT8lppbAHefms9ry1dDFgnuxV+zt7blEu3uDgmTZ5CdHQMn+zZY3akVr3+6lKmz7yDn0yaTLzTGfB5/clhGD59BApbFzi5725g+PCRAKRnZLCroMDkRK2zWt7dRUUMyco2O0a7WGlsvbVVGFFxABgRXTEiYnFX/Kvp387VQlg0RlQ8GCHffFFImAlJm1hpbAFuuXUiSUnJgDWyW+XnrFevNEJDv5kUiIyMNDFN22JjY3l34wYAqk+coGu3biYnEl+zdYFTXlZGYmIiAM54JxUV5SYnap3V8u7fv4/+/QeYHaNdrDS2RpdkvF8exuv14qk+AKGRuI8UUF/4MjhCcEQ7cXTrRUiPwcBXBVFMkml5rTS232WF7Fb5ORuUmcmtE38CwMGDB0jv18/kRK27Y9ZsVq5Yxs/uv5fw8HDS0tLMjmQaw8ePQOHXAufTTz/l6aef9uch2s3r9VpqpXig592xfTsjR11ndoxLEuhjG5o+FnflvqY1NtEJUF9LSM8sQlKuwn1oO15PY/O+jUcKCO0/PmC+n0Af29YEYnYr/pytXLGc+Y8+FnBj+V2f7d3L5MlT+dHNP2bpklcCdnqyMxhffeCmrx6Bwq8Fzpo15s4bp/boQVVVFQA1NdWkpKSalqU9rJS3qqqSkuL97NyZz6HSUkqKi82O1Corja0RGkHYwAmE9hkFjXV4qksISRv21bqcbDzHPgHAU30AI9qJo0uiqXmtNLbfFejZrfZztquggF690uibnm52lDYtX/YG02bMZOpPb2PK1Nt4b1Ou2ZHEx/xW4BQVFZGdbe688bjxE8jP3wFASXExQ4cNMzVPW6yU9yeTJjN6zFiuueZaLuvTh34ZGWZHapWVxtZzpvKbNTeNdRjd0vi68WtExUFIGN7Gc3jOHCckIQOvux531X7T8lppbL8r0LNb6eestraWkpJiho8YgcvlYvu2bWZHalW3uLjmbkOPnj2JiooyOZE5mj6LyrePQOG3Amffvn0MGGDuvPGQrCzqXC4WPbuQ0WPGEhZm3kLM9rBaXpfLxbq1ayjYmc/hw4fNjtMqK42t4QjBXbabxuL3CL1sJKF9rsN9aAfuL4rwnq3GkfgD3KXbcJcVUb/zz5zb+t8Y4TGm5bXS2AKsX7eWrVs3896mXEtkt8rP2ZJXXmbd2tXMvnMG424Yg9PpNDtSq+bccx+vLV3CurVrOHL4MGOvv8HsSOJjhtfr9fr6Rbdv306fPn1oaGhg8eLFLFiwoMU+OTk55OTkAHC8spJ9JYd8HUPEr5wzXjY7QrtVL7vX7Agitjfq2mwKCwvNjtEhCemDuPmpN3z6mp8tvC8gxsEvN/qrrKykoaGBiooKSktLKS4uJuM7rdW5c+cyd+5cALIscAmkiIiIHQXQumCf8kuBM3nyZABKS0v57LPPWhQ3IiIiIv7ktzU4LpeLNWvWkJ8f2PPGIiIiwUyXiXdQVFQUjz76KBs3bqR3797+OoyIiIhIC7b/sE0RERG5sK8vE7cjFTgiIiLSqZYtW4ZhGOTl5fHCCy+wYMEC4uPjSUhIYNasWQAX3NYRtv4sKhEREWldZ6/BOXr0KCdPnmT69OkMGzaMf/zjH0RFRTF//ny2bNlCfX09RUVFLbZ1lAocERGRINbZH7a5evVqhgwZAsDdd9/Nxo0bGTlyJAAZGRkUFBSwYcOGFts6SlNUIiIi4jOVlZXnfVTTt+97B023kKmvr+eDDz7g0KFDNDY2kpjY9Jl6TqeT8vJyysrKWmzrKBU4IiIiQcowwOHjS7sTExNbvZPx6dOnufzyy7nlllv4y1/+whtvfHMnZa/X22Ka60Lb2kNTVCIiItJpunfvTlpaGgC9e/dm9OjRVFVVAVBdXU1qaio9evRosa2jVOCIiIgEMcPw7aMto0ePbu7wVFRUcOutt7Jjxw4AiouLGTZsGBMmTGixraNU4IiIiASxzr6Kavz48Rw5coRVq1bR0NBAVlYWLpeLhQsXMnbsWMLCwi64raO0BkdEREQ6jcPh4De/+c1525588skW+11oW4eO094dP/300+b/XV5eztq1a7/XgUVERMR8nT1F1VnaLHBOnTrF4cOH2bNnD4cPH+bw4cPU1dVd0jXpIiIiIp2hXVNU69ev5+jRo9TV1eH1egkPD+eBBx7wdzYRERHxIwPD55eJB4o2C5yuXbvy0EMPceDAAdLT0zsjk4iIiHSGAJtW8qV2r8H5dnFz7tw5v4QRERER8YV2FzjPPfccS5YsAeDIkSO8+eabfgslIiIinaOzLxPvLO2+TPzyyy9n/PjxwKV/8JVIa7xer9kROqR62b1mR2i3fo+sNjtCh5Q8N8XsCCJice3u4Hz22Wfs3buXxsZGiouL+fzzz/2ZS0RERDqBw8ePQNHuDs7cuXN56aWXWLRoEenp6fzyl7/0Zy4RERGRS9buAicyMpKHH34YALfbzfvvv988ZSUiIiLWY0BArZvxpXYXOH/+85/5/PPPiYyMJDIykurqahU4IiIiFuewZ33T/gKnf//+3HHHHXz88ceMGjWKNWvW+DGWiIiIyKVr93qgkpISTp8+zY4dOzh58iSlpaV+jCUiIiKdwWH49hEoWi1w/vSnP/H+++8DMGHCBE6dOsWsWbN4+umn6d+/f6cEFBEREemoVqeoUlNTufHGG4GmTxC/5pprAPif//kfSkpK/J9ORERE/KbpE8ADqO3iQ60WOOfOneOdd96hS5cufPTRR+d9RMOOHTv4t3/7N78HFBEREf8JpGklX2p1imr69On069cPaLrL7LcfIiIiIoGqzauoBg4cyMCBA0lPT6dXr17N26+66iq/BhMRERH/s+kMVfuvovp2cQMQHx/v8zAiIiIivtDu++CIiIiIvRiAw6YtHBU4IiIiQSyQPiDTl2xf4PzuqQXExcWTkJDAzDtnmR2nTVbKa6Ws0JQ3Pj4epzPw81ppbD3nanHt/wBHTDwe1ymi+o0473nMFeNoPHWcL7f8EUdULABxYx7EER7d6VkPlZby4M/uJ6F7dwBe+FMOXbt27fQcHWGlcwFg76efsmb1Kn79+BNmR2lVTU0Na95cRUREBG63m9l3zzE7kviYXQs3AHYXFREZFcXDj8wnb+sW6uvrzY7UKivltVJWgN27i4iKiuKhefPJywvsvFYb27P7thDV/zqi+l6DI6JLi+cNJw4BEDt0Os5xv8Q57pemFDdfe/yJBbz6+jJefX1ZwBc3VjsXANavW4Pb7TY7Rpu2fZBHt7g47pg1m7ytW8yOY6qme+H47hEobF3g5L67geHDRwKQnpHBroICkxO1zkp5rZQVIHfjBoaPaMrbr18GhbsCN6/VxtYRFsW5w7sB8Jw7jfdc7XnPHRExZsazNKudC7uLihiSlW12jHaZ+JNJTJn6UwDCw8NNTiP+YOsCp7ysjMTERACc8U4qKspNTtQ6K+W1UlaA8vIyundvyhvvdFJRHrh5rTa2UQPG4CreRs3m5zEcoXQZ8tPznod0aZoOOnfkI878cx2ndr5uat7338vl2T88w2+eeNzUHO1htXNh//599O8/wOwY7XbmzBl+MX8ek78qdIKRYRg4fPwIFH4rcEpLS7npppuYMWMGM2bM4NSpU/46VLt4vV5L3Y7aSnmtlBWsldcKWRtrjhKZfg2RvYdw9rO/01hz5LznXncDjqhuRP/gJrr88CdghNB46rgpWROTkphz7/3M/8VjhIaGcshCHxoc6OfCju3bGTnqOrNjdEhsbCx/eHYR77z9FsePm3NOiv/4tYOzYMECli1bxrJl5sx1p/boQVVVFQA1NdWkpKR2eoaOsFJeK2UFSE3twYkTX+WtriYlNXDzWm1sXfs/ICrjOqL6jSCy33C+/Pui856fO/JP8DQ2r7sJ6ZKAx/WlKVnr6+ub34t69uzFsWPHTMnRXlY6F6qqKikp3s/OnfkcKi2lpLjY7Eitqqmpaf7De9CgTD7I22pyIvNoDY4FjRs/gfz8HQCUFBczdNgwkxO1zkp5rZQVYNyPJpD/4Vd5S4rJHhq4ea02tkZETPO7WkhMApEZI897boSGc/azv1Nf9gkAntpqQmKTTMn66pLFbPsgD2iatuzTt68pOdrLSufCTyZNZvSYsVxzzbVc1qcP/TIyzI7UqtdfXcrGDe8AcOxYBX37ppucyDwOw7ePQOHXAic3N5dnnnmGxx9vOdedk5NDdnY22dnZVFZV+uX4Q7KyqHO5WPTsQkaPGUtYWJhfjuMrVsprpawAQ4Zk4XK5eP65hYweHdh5rTa2MT+4ibOfb8Z1cCfu05XE/GDcec/De15JVP/rcLtO4ir5EEd0HCHRcaZknTZjJsePHWP1m6tISkomKcmcQqu9rHYuuFwu1q1dQ8HOfA4fPmx2nFbdPn0GVZWVrPrbSuLi4hmSlWV2JPExw+unT848e/YsVVVV9O7dm9/85jfcc8899OnT54L7ZmVls31noT9iiIVY7UNcA3k9xHf1e2S12RE6pOS5KWZHEOmwUddmU1hord9lPQdcyc9e8O37w7pfzwiIcfBbB+fbc929egX+XLeIiIjYh98KnMWLF5OX1zTXXVZWRt8An+sWEREJRlpk3EEzZ87k2LFjrFq1iuTkwJ/rFhERCTo+XmAcSIuM/fZZVMnJyTzwwAP+enkRERGRi7L9h22KiIjIxRkEUNvFh2x9HxwREREJTurgiIiIBCmDwFo340sqcERERIKYXQscTVGJiIiI7aiDIyIiEsSsdFf2jlAHR0RERGxHHRwREZEgZedFxurgiIiIiO2ogyMiIhKsAuzzo3xJBY6IiEgQc9i0wtEUlYiIiNiOOjgiIiJBSouMRURERCxEHRwREZEgZtMlOCpwREREgpeBA3tWOAFR4HgBr9drdox2sestrQOB1cbW47HGOQtQ/OxksyN0SM/7/mp2hHb74i8zzY4gIhcQEAWOiIiIdD4D+05RaZGxiIiI2I46OCIiIsHKsO9l4ipwREREgpjuZCwiIiJiEergiIiIBCktMhYRERGxEHVwREREgpjW4IiIiIhYhDo4IiIiQcymDRwVOCIiIsHKwL5TOXb9vkRERCSIqYMjIiISrAzrfdBxe9m6wHG73bz26hLi4uL59JM9/PrxJ8yO1KbfPbWAuLh4EhISmHnnLLPjtMpKWcFaef/z6aaszoQEZt4R2FkB/rZiObdNm252jDZ5zp3h3P7NhHRNwetuJKzHldQf3oXhCAWvh4iMMbjPVFK7IwdHZCwAMdfejxEebVpmq5y3breb15YuIS5e77cSGGw9RfXeply6dYtj0uQpxMTE8MmePWZHatXuoiIio6J4+JH55G3dQn19vdmRLspKWcFaeXfvbsr60Lz5fJAX2FkB3n5rPa++usTsGO1ybv/fCe9zLeG9h+I5fYyGo0U4wqOJ6HcdDcf2Nu8X9cOf0mX0I3QZ/YipxY2Vztv3NuXSLa7p/TY6Wu+3VmL4+BEobF3g9OqVRmjoN02qyMhIE9O0LffdDQwfPhKA9IwMdhUUmJzo4qyUFayVd9O3s/bLoHBX4GYFuOXWiSQnJZsdo108ZypxRMUDYER1w4jsRljvYU3PHSFmRrsgK523er+1JoOm++D48hEobD1FNSgzk0GZmQAcPHiA9H79TE7UuvKyMhITEwFwxjupqCg3OdHFWSkrWCtveVkZ3btbI6vVOGJTcNeUEpLQD3flfozUTGiso7ZoeXOhA9BQ/i8aT5TgPXeG6KvNm3qz0nmr91vpqE8//ZRVq1bxxBNPsGDBAuLjm6YLZ81qmi680LaO8GsHZ9myZSxfvpyHHnrIn4dp099WLOeR+Y9ZaiGV1+u1TF4rZQVr5bVSViuIHHAjDRV7qS/ZiqNLIkZ4NEZYFDHXzKHh6G48rpM4IrsSkXE9UVfcAkYI7jOVZscGrHMurFyxnPmP6v3WSsyaolqzZg1ut5uioiKioqKYP38+W7Y0TRdeaFtH+a3AOXr0KCdPnmT69OkMGzYMr9frr0O1ateuAnqlpdE3Pd2U43dEao8eVFVVAVBTU01KSqrJiS7OSlnBWnlTe/TgxImmrNUBntVqjLBIojInEpExFm9DHY6IWLz1ZwEIiU+j8dhevJ5GjPAoABzRTryuk6bltdJ5C7CroIBevfR+K20rKioiOzsbgA0bNjByZNN0YUZGBgUFBRfc1lF+K3BWr17NkCFDALj77rtNqY5ra2s5UFzMtcNH4HK52LF9W6dn6Ihx4yeQn78DgJLiYoYOG9bGV5jHSlnBWnnHjZ9A/odNWQ+UFJM9NHCzWo379DHqv/gnAN4GF42V+6j/4qOm564vccQmUV+cR+Oxz77aVoOjS6JZcS113tbW1lJSUszwEU3vt9u36f3WKgzDt4/22LdvHwMGDACg7NvThU4n5eXlF9zWUX4rcEpLS8nLy+OZZ55h/vz5pnRwli5+mXVr13DXnTMZf+NY4uOdnZ6hI4ZkZVHncrHo2YWMHjOWsLAwsyNdlJWygrXyXj0ki7o6F88/t5DrRgd2VoD169aydetm3tuUa3aUNhmOUBoO76Ju70YiMkYT3mcE3rrT1JfmY4THEJqQTnjfEXhcJ6k/VIAR2Q1HVDfT8lrpvF3yysusW7ua2XfOYNwNY3A69X5rDQaG4dtHZWUl2dnZzY+cnJzzjrh9+3auu+66C6a50HThpU4hGl4/VR5z585l0qRJ3HLLLfzlL39hwIAB531DOTk5zd/08cpKPi8u9UcMnwvmeVo5n8djzrTrpbDaadvr/mVmR2i3L/4y0+wIEiBGXZtNYWGh2TE6JP2KH/Kfr7/j09f8/QOTWh2HNWvWEBcXR0VFBRs3bqRv377cdNNNjBw5kv/zf/4Po0ePZsuWLYwdO/a8bV9PWbWX3zo43bt3Jy0tDYDevXtTVlZ23r/PnTuXwsJCCgsLm68YERERkc7z9WdR+fLRlsmTJzN27FiuvfZa+vTpw6233sqOHU3ThcXFxQwbNowJEya02NZRfitwRo8e3VzBVVRUMHDgQH8dSkRERCzE5XKxZs0a8vPzSUxMxOVysXDhQsaObZouzMrKarGto/x2H5zx48fz9NNPs2rVKhoaGhg8eLC/DiUiIiKXyIylF1FRUTz66KM8+uijADz55JMt9rnQto7wW4HjcDj4zW9+46+XFxEREbkoW9/JWERERFpnsWsQ2k0FjoiISLAy7Ht1sK0/bFNERESCkzo4IiIiQerry8TtyK7fl4iIiAQxdXBERESCmF3X4KjAERERCWL2LG80RSUiIiI2pA6OiIhIELPpDJU6OCIiImI/6uCIiIgEqabLxO3ZwlGBIyIiEsQ0RSUiIiJiEergiIiIBC0Dw6ZTVOrgiIiIiO2ogyMiIhLE7LoGJyAKHAPr3Cq60e0xO0KHWGVcAUIc1skK4LBYXiv54i8zzY7Qbol3LjE7QodUvn632RHazePxmh3B9ux8FZWmqERERMR2AqKDIyIiIiYw7DtFpQ6OiIiI2I46OCIiIkFMHRwRERERi1AHR0REJIjZ9UZ/KnBERESClAHY9Y4XmqISERER21EHR0REJIjZdYpKHRwRERGxHXVwREREgphdLxNXgSMiIhLENEUlIiIiYhHq4IiIiAQpO18mbvsC53dPLSAuLp6EhARm3jnL7Dht+q+nf0tcfDxOp5MZdwR+3r+tXM5tt083O0a7rFyxnNunWSOr1c5bK+W1UlZvfS0NB7bi6JKM19NIWO9raDicDxh4Kj8nIuuuC+5jBrfbzWtLlxAXH8+nn+zh148/YUqOjrDS+5d0nK2nqHYXFREZFcXDj8wnb+sW6uvrzY7Uqo92N+X9/x5+hA/ytgZ83nfeWs/rS5eYHaNd3n5rPa8tXWx2jHax2nlrpbxWygrQULKF0LRhhPbKwnvmGJ6z1dDgIqz3NTgS0vF6vS328dbXmpL1vU25dIuLY9LkKURHx/DJnj2m5Givd95az2sWef/yL8Pn/wUKWxc4ue9uYPjwkQCkZ2Swq6DA5ESt2/TuRq4dPgKA9H4ZFO4K7Lw/vnUiScnJZsdol1tunUhSkjWyWu28tVJeK2UF8NRWYkTGAWBEdqPxaCGO+MsACOszCsMwWuzjrj5gStZevdIIDf1mUiAyMtKUHO1lpfcvvzKarqLy5SNQ2HqKqrysjMTERACc8U4qKspNTtS68vIyun+VNz4+nmMBnlf8w3LnrYXyWikrgCM2Gc+Xh3E4++I+UYL7+F7wunFX7sN79gThg+9osU9IaIQpWQdlZjIoMxOAgwcPkN6vnyk5RL7mtw7Oxx9/zNVXX82MGTOYMGECb775pr8O1S5erxcjkErLtlgtr/iF1c5bK+W1QtawfjfgrvyMxtJtOGK6g+HAEZtK+MAJOLr1wlO1r8U+Rli0qZlXrljO/EcfC/ixlW8YPn4ECr8VOCdOnGD79u0sW7aMu+66i0mTJvnrUBeV2qMHVVVVANTUVJOSktrpGToiNbUHJ77KW11TQ3KA5xX/sNx5a6G8VsoKYIRGEH75jwnrex3ehjpCnH0xohOa/i06AY/ryxb7OLqYN+2yq6CAXr3S6JuebloGka/5rcC5/vrriY6O5ty5c7jdbkJCQvx1qIsaN34C+fk7ACgpLmbosGGdnqEjbhr/I3bmfwjAgZJisocGdl7xD6udt1bKa6WsAJ4zx2ks/xgAb2MdoZeNwFN9sOl53UkcsSkt9nHEmlPg1NbWUlJSzPARI3C5XGzfts2UHNIxTZeJGz59BAq/LzJesWIF48aNa7E9JyeH7OxssrOzqayq9Muxh2RlUedysejZhYweM5awsDC/HMdXrh7SlPeFRc9y3egxAZ/3rXVrydu6mfc35ZodpU3r161l69bNvGeBrFY7b62U10pZAXCE0vhFEfX7NhHWZxQhKZl4XdU0Ht0FHjch8Ze12McsS155mXVrVzP7zhmMu2EMTqfTtCzt8fX7lxXeE+TSGF6v1+vPAzz66KMsXLiw1X2ysrLZvrPQnzF8ptHtMTtCh1hpHjzErnebEltLvNNalxpXvn632RHazePx668nnxs9YiiFhdb4Xfa1H1x5Na+s3uzT13x4xk0BMQ5+vYqqrq6OY8eO+fMQIiIi8n3Y9G9Lv05Rff7550REmHPJooiIiAQvv3ZwQkJCAn4eVkREJJgF0t2HfcmvHZzMzEyeeeYZfx5CREREpAVb38lYREREWmeha1E6RAWOiIhIELNpfWPvD9sUERGR4KQOjoiISDCzaQtHHRwRERGxHXVwREREglTTJ4Dbs4WjAkdERCRYGfa9ikpTVCIiImI76uCIiIgEMZs2cNTBEREREftRB0dERCSY2bSFow6OiIiI2I46OCIiIkHL0GXiIiIiYj+6TFxERETEItTBERERCVIGtl1jHBgFjhfwer1mx2iX0BA1vUTkG5Wv3212hA5JmfOa2RHarfyVO82OIBYWEAWOiIiImMSmLRwVOCIiIkHMrldRab5FREREbEcdHBERkSCmy8RFRERELEIdHBERkSBm0waOChwREZGgZcKNcNxuN0uWLCE+Pp49e/bwxBNPsGDBAuLj40lISGDWrFkAF9zWEZqiEhERkU6Tm5tLXFwcU6ZMISYmhry8PKKiopg/fz5btmyhvr6eoqKiFts6SgWOiIhIEDN8/F9b0tLSCA39ZgJp8+bNjBw5EoCMjAwKCgrYsGFDi20dpSkqERER8ZnKykqys7Obn8+dO5e5c+c2P8/MzCQzMxOAAwcO4PV6SUxMBMDpdFJeXk5ZWVmLbR2lAkdERCRIGfj+MvHExEQKCwvb3G/58uU89thj/P73v2/e5vV6Mb4T6ELb2kNTVCIiItKpCgoKSEtLIz09nR49elBVVQVAdXU1qampF9zWUSpwREREgpjh40dbamtrKS4uZsSIEbhcLkaNGsWOHTsAKC4uZtiwYUyYMKHFto6y/RTV755quszM6Uxg5p0dv8yss/3uqQXExTVdFhfoea2UFWDliuXcPm262THaxWpja6W8VsoK1jlvvedqqS/ZgiM2Ga+7kfA+1wLgPvkFDYd3EXnlZADqSz8EDNzHPyNq2BzT8oL1fj/4TSdfJv7yyy+Tl5fHunXrOHDgAIsXL8blcrFw4ULGjh1LWFgYWVlZvP322+dt6yhbd3B27266zOyhefPJy7u0y8w60+6iIiKjonj4kfnkbQ3svFbKCvD2W+t5belis2O0i9XG1kp5rZQVrHXe1hdvJqz3NYSlZeM5cwxvfS0ADUf+AV4PAJ6z1XgbXIT3uZaQ7v3wer2m5bXa7wc7mTdvHitXrmTZsmUUFBRwxRVX8OSTT/Loo48ye/bs5v0utK0jbF3g5G7cwPARTZeZ9euXQeGujl9m1ply393A8OFNedMzMth1CZfFdRYrZQW45daJJCUlmx2jXaw2tlbKa6WsYK3z1nOmEiMqDgBHZDcaq0pwV5cSktC3eZ+GI4WEOPsAEJ5+3SUtHPUVq/1+8KfOvky8s9i6wCkvL6N796bLzOKdTiou4TKzzlT+7cvi4p1UVARuXitltRqrja2V8lopq9U4uibjrjmE1+ulsaoYb30t7lPlhMSmNO/jOVOF+/hnnNu7AVfhq6Z2cKz2+0E6zm9rcE6ePMkbb7xBamoqlZWVPPDAA/46VLtc6mVmZrFSXitltRqrja2V8lopqxWE97+B+n3v4zn5BY6Y7njOHCc8/TrwuJv38Ta6cHS9grCeg6kv3oK7ch+hSQNNTP1VriA/F+z6rfutg7N06VLuuOMOJk+ejNPpZM+ePf461EWlpvbgxImmy8xqqqtJuYTLzDpT6rcui6upqSYlJXDzWimr1VhtbK2U10pZrcYIjSTiilsI7zcab2MdhiMUz+ljNFYV46mtxH36GI6IWBzRzqb9YxLwnK0xLa/Vfj/4U2dfRdVZ/FbgxMbGsmHDBgBOnDhBt27d/HWoixr3ownkf9h0mVlJSTHZQzt+mVlnGjd+Avn5X+UtLmboJVwW11mslNVqrDa2VsprpaxW4zl9nIayj5ueNNQRccUthCb/gNDuGThiEgmJTSYk6XLc1QcB8LpOEtLVvKLCar8fpOP8VuDMnj2bZcuWce+99xIeHk5aWpq/DnVRQ4Zk4XK5eP65hYwefWmXmXWmIVlZ1LlcLHp2IaPHBHZeK2UFWL9uLVu3bua9TblmR2mT1cbWSnmtlBWsdd4SEkrjkULOffYuYenXAeBtrKfh6D9wVxXjqa0iNDUTT+0JGg7vAq+bEOdlpsW12u8Hv7JpC8fw+mmV1549eygqKiI6OppFixaxadMmwsPDm/89JyeHnJwcAI5XVvJ5cak/YvhcMM/Tioj1pcx5zewI7Vb+yp1mR+iQ64YPbddHFASSzB8O4c3cbT59zTtuGR0Q4+C3RcZvvPEGv/3tbwkLC6O8vJzc3FxuvfXW5n//9odvDcnKvtjLiIiIiJ80NV3s+Ye73wqcuLi45m5Hz549iYqK8tehRERE5FIY9r2Kym8Fzn333ceSJUtISEjg8OHDzJ8/31+HEhERETmP3wqchIQE7rvvPn+9vIiIiPiATRs49r6TsYiIiAQn23+auIiIiLTCpi0cdXBERETEdtTBERERCVqB9QngvqQCR0REJIjZ9TJxTVGJiIiI7aiDIyIiEqQC7OOjfEodHBEREbEddXBERESCmU1bOCpwREREgphdr6LSFJWIiIjYjjo4IiIiQUyXiYuIiIhYhDo4IiIiQcymDRwVOCIiIkHL0BSViIiIiGUERAfHAAyLlJBer9fsCB1ilXG1IiudCzoP5GsVi2eZHaHd4qf9xewIHTLI7ACXzJ7vD+rgiIiIiO0ERAdHREREOl/TDIrZKfxDHRwRERGxHXVwREREgphNGzgqcERERIKZpqhERERELEIdHBERkSCmTxMXERERsQh1cERERIKZPRs4KnBERESCmU3rG01RiYiIiP2ogyMiIhKkDH2auIiIiIh1qIMjIiISxOx6mXhQFDgrVyzn9mnTzY7RLr97agHx8fE4nQnMvHOW2XFa9bunFhAXF09CQuBnBWvltdJ5ANYbW6tkramp4ZWXX6Jfvwzq6+sD/n1sxfJlGIbB9g/yWLjoBbPjtMpbX4u79AOMLkngacSRNOi85yG9hgHQsHctRlgMRngMIb2Hm5zaT+xZ39h/iurtt9bz2tLFZsdol927i4iKiuKhefPJy9tCfX292ZEuandREZFRUTz8yHzytgZ2VrBWXiudB2CxsbVQVoC/vJTDT2+bxqTJUzhQUsyXX35pdqSLOnr0KKdOnuT2adPJHjoMr9drdqRWuUvzCOk1lJAeQ/DWHm/x3Ft/Fs+XhzBCwgnNuAlP1ed4PY1mx5YOsH2Bc8utE0lKSjY7RrvkbtzA8BEjAejXL4PCXQUmJ7q43Hc3MHx4U9b0jAx2FQRuVrBWXiudB2CxsbVQVoBDBw+SmpoKQFJSMrsKdpqc6OLWrVnN4KuHADDrrrsxAnzlqre2CiK7NT2J6Ir74NbznntqDuA59i8czgwAjJgkvNUHTUrrX4aPH4HC9gWOlZSXl9G9eyIA8U4nFeXlJie6uPKyMhITm7I6451UVARuVrBWXiudB2CxsbVQVoB+/fvzz48+wuv1UrAzn+rqarMjXdShQ6Vs25bHs394hl/+Yn7Ad3CMLkl4vzyM1+vFW32AkD6jzntOfS1e10mIiG36gvAYvOdOmhtaOsRva3Bqamp46aWXyMhomjuePj2w544DjdfrDfi/gL5mpaxgrbxWygrWymuFrA/M/Tkv/vmP7P30Ey7r04e4uDizI13UmdOnGTjwcm7+8S0sfvkvbN++jVGjrjM71kWF9B2L+8BmOF2GEd0do1tvPFX7mp8TFm12xE4T4D8Gl8xvHZycnBymTZvGlClTKC4O7LnjQJGa2oMTJ6oAqKmuJuWr1nQgSu3Rg6qqr7LWVJOSErhZwVp5rXQegMXG1kJZAWJiYnjsl7/irjn3cPr0aTIy+psd6aISunenV680ANJ696a8rMzkRK0zQiMIHTCB0MtG4W2sw+iS/J3nSRhR3aD+dNMX1NdifD2FJZbgtwLn4LfmjpOTk9m5M3DnjgPFuB9NIP/DHQCUlBSTPXSYyYkubtz4CeTnf5W1uJihwwI3K1grr5XOA7DY2FooK8DBAwfIfXcj0NQh6ZeRYXKiixt13WiK/lEIwLGKCgYMGGhyotZ5aitxH9vT9KSxDgzHec8dXZJxJGXiOVECgLf2OEZ8X5PS+pPh8/8Chd8KnP79+/PRV3PH+fkt545zcnLIzs4mOzubyqpKf8Vg/bq1bN26mfc25frtGL4yZEgWLpeL559byOjRYwkLCzM70kUNycqizuVi0bMLGT0msLOCtfJa6TwAi42thbIChIWHs27Nal5Y9Byz7ppjdpxW3TRuPEePHmH1m6toaGjgh4MHmx2pVYYjBE9ZEY0l7xHSe0SL5wCO+D7gaaCxeBOO7gMxHPa7s4rBN3cz9tUjUBheP60Eq62t5Y9//CPdu3fniy++ICsri5tvvvmC+2ZlZbN9Z6E/YvhcoC+c+65AX2NgZVY6F3QeiBXFT/uL2RE6ZNDhP1FYaI3fZV+7ekg2f9/m2xmWcaOvCYhx8FsHJyYmhl/96lfcc0/T3HH//oE7dywiIiL24rcC58CBA2zc2DR33LQ4LnDnjkVERMRe/DahGB4ezurVq9m3bx9z5szx12FERETke7DrDLbfCpxevXrx4osv+uvlRURExAcC6conX9KdjEVERMR27HfNm4iIiLRPgF3a7Uvq4IiIiIjtqIMjIiISpALtE8B9SQWOiIhIMLNphaMpKhEREbEddXBERESCmC4TFxEREbEIdXBERESCmC4TFxEREbEIdXBERESCmE0bOCpwREREgppNKxwVOCIiItKpFixYQHx8PAkJCcyaNcsvx9AaHBERkSBm+Pi/thQVFREVFcX8+fPZsmUL9fX1fvm+VOCIiIhIp9mwYQMjR44EICMjg4KCAr8cR1NUIiIiQcqg8y8TLysrIzExEQCn00l5eblfjhMQBc6hQ6WMujbb569bWVnZPIiBzkpZwVp5rZQVrJXXSlnBWnmtlBX8k3eQT1/tG/4a29LSUp+/pr8lJnb3+e9fl8tFdvY3rzl37lzmzp17wX29Xi+GnyqsgChwqqqq/PK62dnZFBYW+uW1fc1KWcFaea2UFayV10pZwVp5rZQVrJXXSln9bePGjZ1+zB49elBVVcXAgQOprq4mMzPTL8fRGhwRERHpNBMmTGDHjh0AFBcXM2zYML8cRwWOiIiIdJqsrCxcLhcLFy5k7NixhIWF+eU4ATFF5S8Xm/MLRFbKCtbKa6WsYK28VsoK1sprpaxgrbxWympXTz75pN+PYXi9Xq/fjyIiIiLSiTRFJSIiIrZj2ymqzrgNtC8tX76c6dOnmx2jTW63myVLlhAfH8+ePXt44oknzI50UTU1NaxatYqIiAjcbjdz5swxO1K7fPrpp6xatSqgx7a0tJT777+f7t27A5CTk0PXrl1NTtW6ZcuWYRgGeXl5vPDCC2bHuaiPP/6Yu+++m4EDB/Lll18yd+5cpk6danasizp58iRvvPEGqampVFZW8sADD5gd6aJqamp46aWXyMjIoL6+3hLvuXLpbNnB6azbQPvK+vXrWbx4sdkx2iU3N5e4uDimTJlCTEwMe/bsMTvSReXl5REXF8fs2bPZsmWL2XHabc2aNbjdbrNjtGnBggUsW7aMZcuWBXxxc/ToUU6ePMn06dMZNmwYgTwzf+LECbZv386yZcu46667mDRpktmRWrV06VLuuOMOJk+ejNPpDOj3hJycHKZNm8aUKVMoLi7myy+/NDuS+JEtC5zOug20r0ycOJHk5GSzY7RLWloaoaHfNP4iIyNNTNO6SZMm8dOf/hSA8PBwk9O0T1FR0Xk3yBLfWL16NUOGDAHg7rvv9tuNxXzh+uuvJzo6mnPnzuF2uwkJCTE7UqtiY2PZsGED0FScdevWzeREF3fw4EFSU1MBSE5OZufOnSYnEn+yZYHTWbeBDkaZmZn85Cc/AeDAgQP069fP5EStO3PmDPPmzWsudALdvn37GDBggNkx2iU3N5dnnnmGxx9/3OwobSotLSUvL49nnnmG+fPnB3QH52srVqxg3LhxZsdo0+zZs1m2bBn33nsv4eHhpKWlmR3povr3789HH32E1+slPz+f6upqsyOJH9mywPk2f94GOpgtX76cxx57LODHNjY2lkWLFvHWW29x/Phxs+O0avv27Vx33XVmx2iXpKQk7r//fh577DFCQ0MD/hb1p0+f5vLLL+exxx7jqquuYtu2bWZHatM//vEPUlJSzI7Rpr179zJ16lR+/OMf88orrwT0koCf//znbN26lcWLF9OnTx/i4uLMjiR+ZMsC5+vbQANUV1c3tyTFNwoKCkhLSyM9Pd3sKK2qqanh1KlTQFPnaevWrSYnal1lZSX79+8nPz+f0tJSiouLzY50UfX19c3rbnr16sWxY8dMTtS67t27N3cWevfuTVlZmcmJWldXVxfwY/q1N954g5kzZ3Lbbbdx2223kZuba3aki4qJieFXv/oV99xzD6dPn6Z///5mRxI/smWB01m3gQ5GtbW1FBcXM2LECFwuV0D/Jbx06VLeeecdACoqKgK+IJs8eTJjx47l2muvpU+fPmRkZJgd6aIWL15MXl4e0DQl3LdvX5MTtW706NHNnz1UUVHBwIEDTU7Uus8//5yIiAizY7RLXFxccye3Z8+eREVFmZzo4g4cOND82UunT58O6J8x+f5sWeB01m2gfWXt2rVs3rw5oP/y+drLL7/M6tWrmTFjBmPGjMHpdJod6aJmzJhBZWUlK1euJD4+nqysLLMjtcnlcrFmzRry8/M5fPiw2XEuaubMmRw7doxVq1aRnJxMUlKS2ZFaNX78eI4cOcKqVatoaGhg8ODBZkdqVUhISED/bH3bfffdx5IlS1izZg2HDx/mhhtuMDvSRYWHh7N69Wqee+45y9w2Qi6d7mQsIiIitmPLDo6IiIgENxU4IiIiYjsqcERERMR2VOCIiIiI7ajAEREREdtRgSMiIiK2owJHREREbEcFjoiIiNhOqNkBROTS5eTkkJyczIcffkjPnj258sorGT58OHfddRfLly8H4KWXXsLpdFJSUsKvfvUrPB4Pf/rTn+jRowd//vOfeffddzlz5gxLly6le/furFq1qvlrRUSsSh0cEYv65z//Sc+ePZk0aRKRkZFceeWVAERERPCDH/wAgI8//piQkBCmTp1KSkoK+fn5nD59mpKSEiZPnswzzzwDwMGDBzl37hzTpk3jySefNO17EhHxFRU4IhbldDqpqakBaP5k7+/au3cv1dXVbNmyhdDQUBwOB926dePGG29k1qxZ/Otf/wLgyiuvJCEhgTvvvDPgP2lbRKQ9VOCIWFRMTAyHDh1i06ZNTJkyhYiICBobGwGorq4GoF+/fjidTsaOHcvUqVPJyMhg7969XH/99bz++uvs3r2bU6dOUVRUxF133cVrr73Gm2++aea3JSLiE1qDI2JRDQ0N7Nixg4qKCnr16sXDDz/Mo48+yuHDhzl37hxFRUVkZ2eTl5fH0qVLAbjjjjsoLS3l/vvv56677qJPnz7ExsZSWlrK0qVLufnmmxk6dKjJ35mIyPenTxMXsaiXXnqJiRMnkpyczNatWzlx4gRTp041O5aISEBQgSNiUZ9++ikFBQX07t2bw4cPc+utt9K9e3ezY4mIBAQVOCIiImI7WmQsIiIitqMCR0RERGxHBY6IiIjYjgocERERsR0VOCIiImI7KnBERETEdv5/ctBURc1YU80AAAAASUVORK5CYII=\n",
      "text/plain": [
       "<Figure size 1152x576 with 2 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# 输出混淆矩阵\n",
    "plt.figure(figsize=(16,8), facecolor='w')\n",
    "\n",
    "from pylab import mpl\n",
    "mpl.rcParams['font.sans-serif'] = ['FangSong']  # 指定默认字体\n",
    "mpl.rcParams['axes.unicode_minus'] = False  # 解决保存图像是负号'-'显示为方块的问题\n",
    "plot_confusion_matrix(np.argmax(y_test, axis=1),y_pred,\n",
    "                      title='测试集混淆矩阵')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0fd1ab4c",
   "metadata": {},
   "source": [
    "加分项 1，保存模型并加载模型，用自己手写的数字图片（手机的便签或微 软画图工具，保证没有阴影）测试分类效果，像素需要降低尽量保持在 28x28x1。"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "35a5d84f",
   "metadata": {},
   "source": [
    "（1）保存模型并加载模型"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "adb5c8e8",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "ERROR:root:Internal Python error in the inspect module.\n",
      "Below is the traceback from this internal error.\n",
      "\n",
      "ERROR:root:Internal Python error in the inspect module.\n",
      "Below is the traceback from this internal error.\n",
      "\n",
      "ERROR:root:Internal Python error in the inspect module.\n",
      "Below is the traceback from this internal error.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\29459\\anaconda3\\envs\\python37\\lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 3437, in run_code\n",
      "    exec(code_obj, self.user_global_ns, self.user_ns)\n",
      "  File \"<ipython-input-22-08ba37669687>\", line 1, in <module>\n",
      "    model.save('mnist_cnn.h5')\n",
      "  File \"C:\\Users\\29459\\anaconda3\\envs\\python37\\lib\\site-packages\\keras\\engine\\network.py\", line 1152, in save\n",
      "    save_model(self, filepath, overwrite, include_optimizer)\n",
      "  File \"C:\\Users\\29459\\anaconda3\\envs\\python37\\lib\\site-packages\\keras\\engine\\saving.py\", line 449, in save_wrapper\n",
      "    save_function(obj, filepath, overwrite, *args, **kwargs)\n",
      "  File \"C:\\Users\\29459\\anaconda3\\envs\\python37\\lib\\site-packages\\keras\\engine\\saving.py\", line 540, in save_model\n",
      "    with H5Dict(filepath, mode='w') as h5dict:\n",
      "  File \"C:\\Users\\29459\\anaconda3\\envs\\python37\\lib\\site-packages\\keras\\utils\\io_utils.py\", line 191, in __init__\n",
      "    self.data = h5py.File(path, mode=mode)\n",
      "  File \"C:\\Users\\29459\\AppData\\Roaming\\Python\\Python37\\site-packages\\h5py\\_hl\\files.py\", line 408, in __init__\n",
      "    swmr=swmr)\n",
      "  File \"C:\\Users\\29459\\AppData\\Roaming\\Python\\Python37\\site-packages\\h5py\\_hl\\files.py\", line 179, in make_fid\n",
      "    fid = h5f.create(name, h5f.ACC_TRUNC, fapl=fapl, fcpl=fcpl)\n",
      "  File \"h5py\\_objects.pyx\", line 54, in h5py._objects.with_phil.wrapper\n",
      "  File \"h5py\\_objects.pyx\", line 55, in h5py._objects.with_phil.wrapper\n",
      "  File \"h5py\\h5f.pyx\", line 108, in h5py.h5f.create\n",
      "OSError: Unable to create file (unable to truncate a file which is already open)\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\29459\\anaconda3\\envs\\python37\\lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 2061, in showtraceback\n",
      "    stb = value._render_traceback_()\n",
      "AttributeError: 'OSError' object has no attribute '_render_traceback_'\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\29459\\anaconda3\\envs\\python37\\lib\\site-packages\\IPython\\core\\ultratb.py\", line 1101, in get_records\n",
      "    return _fixed_getinnerframes(etb, number_of_lines_of_context, tb_offset)\n",
      "  File \"C:\\Users\\29459\\anaconda3\\envs\\python37\\lib\\site-packages\\IPython\\core\\ultratb.py\", line 248, in wrapped\n",
      "    return f(*args, **kwargs)\n",
      "  File \"C:\\Users\\29459\\anaconda3\\envs\\python37\\lib\\site-packages\\IPython\\core\\ultratb.py\", line 281, in _fixed_getinnerframes\n",
      "    records = fix_frame_records_filenames(inspect.getinnerframes(etb, context))\n",
      "  File \"C:\\Users\\29459\\anaconda3\\envs\\python37\\lib\\inspect.py\", line 1502, in getinnerframes\n",
      "    frameinfo = (tb.tb_frame,) + getframeinfo(tb, context)\n",
      "  File \"C:\\Users\\29459\\anaconda3\\envs\\python37\\lib\\inspect.py\", line 1460, in getframeinfo\n",
      "    filename = getsourcefile(frame) or getfile(frame)\n",
      "  File \"C:\\Users\\29459\\anaconda3\\envs\\python37\\lib\\inspect.py\", line 696, in getsourcefile\n",
      "    if getattr(getmodule(object, filename), '__loader__', None) is not None:\n",
      "  File \"C:\\Users\\29459\\anaconda3\\envs\\python37\\lib\\inspect.py\", line 733, in getmodule\n",
      "    if ismodule(module) and hasattr(module, '__file__'):\n",
      "  File \"C:\\Users\\29459\\anaconda3\\envs\\python37\\lib\\site-packages\\tensorflow\\__init__.py\", line 50, in __getattr__\n",
      "    module = self._load()\n",
      "  File \"C:\\Users\\29459\\anaconda3\\envs\\python37\\lib\\site-packages\\tensorflow\\__init__.py\", line 44, in _load\n",
      "    module = _importlib.import_module(self.__name__)\n",
      "  File \"C:\\Users\\29459\\anaconda3\\envs\\python37\\lib\\importlib\\__init__.py\", line 127, in import_module\n",
      "    return _bootstrap._gcd_import(name[level:], package, level)\n",
      "  File \"<frozen importlib._bootstrap>\", line 1006, in _gcd_import\n",
      "  File \"<frozen importlib._bootstrap>\", line 983, in _find_and_load\n",
      "  File \"<frozen importlib._bootstrap>\", line 965, in _find_and_load_unlocked\n",
      "ModuleNotFoundError: No module named 'tensorflow_core.estimator'\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\29459\\anaconda3\\envs\\python37\\lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 3437, in run_code\n",
      "    exec(code_obj, self.user_global_ns, self.user_ns)\n",
      "  File \"<ipython-input-22-08ba37669687>\", line 1, in <module>\n",
      "    model.save('mnist_cnn.h5')\n",
      "  File \"C:\\Users\\29459\\anaconda3\\envs\\python37\\lib\\site-packages\\keras\\engine\\network.py\", line 1152, in save\n",
      "    save_model(self, filepath, overwrite, include_optimizer)\n",
      "  File \"C:\\Users\\29459\\anaconda3\\envs\\python37\\lib\\site-packages\\keras\\engine\\saving.py\", line 449, in save_wrapper\n",
      "    save_function(obj, filepath, overwrite, *args, **kwargs)\n",
      "  File \"C:\\Users\\29459\\anaconda3\\envs\\python37\\lib\\site-packages\\keras\\engine\\saving.py\", line 540, in save_model\n",
      "    with H5Dict(filepath, mode='w') as h5dict:\n",
      "  File \"C:\\Users\\29459\\anaconda3\\envs\\python37\\lib\\site-packages\\keras\\utils\\io_utils.py\", line 191, in __init__\n",
      "    self.data = h5py.File(path, mode=mode)\n",
      "  File \"C:\\Users\\29459\\AppData\\Roaming\\Python\\Python37\\site-packages\\h5py\\_hl\\files.py\", line 408, in __init__\n",
      "    swmr=swmr)\n",
      "  File \"C:\\Users\\29459\\AppData\\Roaming\\Python\\Python37\\site-packages\\h5py\\_hl\\files.py\", line 179, in make_fid\n",
      "    fid = h5f.create(name, h5f.ACC_TRUNC, fapl=fapl, fcpl=fcpl)\n",
      "  File \"h5py\\_objects.pyx\", line 54, in h5py._objects.with_phil.wrapper\n",
      "  File \"h5py\\_objects.pyx\", line 55, in h5py._objects.with_phil.wrapper\n",
      "  File \"h5py\\h5f.pyx\", line 108, in h5py.h5f.create\n",
      "OSError: Unable to create file (unable to truncate a file which is already open)\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\29459\\anaconda3\\envs\\python37\\lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 2061, in showtraceback\n",
      "    stb = value._render_traceback_()\n",
      "AttributeError: 'OSError' object has no attribute '_render_traceback_'\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\29459\\anaconda3\\envs\\python37\\lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 3357, in run_ast_nodes\n",
      "    if (await self.run_code(code, result,  async_=asy)):\n",
      "  File \"C:\\Users\\29459\\anaconda3\\envs\\python37\\lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 3454, in run_code\n",
      "    self.showtraceback(running_compiled_code=True)\n",
      "  File \"C:\\Users\\29459\\anaconda3\\envs\\python37\\lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 2064, in showtraceback\n",
      "    value, tb, tb_offset=tb_offset)\n",
      "  File \"C:\\Users\\29459\\anaconda3\\envs\\python37\\lib\\site-packages\\IPython\\core\\ultratb.py\", line 1368, in structured_traceback\n",
      "    self, etype, value, tb, tb_offset, number_of_lines_of_context)\n",
      "  File \"C:\\Users\\29459\\anaconda3\\envs\\python37\\lib\\site-packages\\IPython\\core\\ultratb.py\", line 1268, in structured_traceback\n",
      "    self, etype, value, tb, tb_offset, number_of_lines_of_context\n",
      "  File \"C:\\Users\\29459\\anaconda3\\envs\\python37\\lib\\site-packages\\IPython\\core\\ultratb.py\", line 1125, in structured_traceback\n",
      "    tb_offset)\n",
      "  File \"C:\\Users\\29459\\anaconda3\\envs\\python37\\lib\\site-packages\\IPython\\core\\ultratb.py\", line 1082, in format_exception_as_a_whole\n",
      "    last_unique, recursion_repeat = find_recursion(orig_etype, evalue, records)\n",
      "  File \"C:\\Users\\29459\\anaconda3\\envs\\python37\\lib\\site-packages\\IPython\\core\\ultratb.py\", line 382, in find_recursion\n",
      "    return len(records), 0\n",
      "TypeError: object of type 'NoneType' has no len()\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\29459\\anaconda3\\envs\\python37\\lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 2061, in showtraceback\n",
      "    stb = value._render_traceback_()\n",
      "AttributeError: 'TypeError' object has no attribute '_render_traceback_'\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\29459\\anaconda3\\envs\\python37\\lib\\site-packages\\IPython\\core\\ultratb.py\", line 1101, in get_records\n",
      "    return _fixed_getinnerframes(etb, number_of_lines_of_context, tb_offset)\n",
      "  File \"C:\\Users\\29459\\anaconda3\\envs\\python37\\lib\\site-packages\\IPython\\core\\ultratb.py\", line 248, in wrapped\n",
      "    return f(*args, **kwargs)\n",
      "  File \"C:\\Users\\29459\\anaconda3\\envs\\python37\\lib\\site-packages\\IPython\\core\\ultratb.py\", line 281, in _fixed_getinnerframes\n",
      "    records = fix_frame_records_filenames(inspect.getinnerframes(etb, context))\n",
      "  File \"C:\\Users\\29459\\anaconda3\\envs\\python37\\lib\\inspect.py\", line 1502, in getinnerframes\n",
      "    frameinfo = (tb.tb_frame,) + getframeinfo(tb, context)\n",
      "  File \"C:\\Users\\29459\\anaconda3\\envs\\python37\\lib\\inspect.py\", line 1460, in getframeinfo\n",
      "    filename = getsourcefile(frame) or getfile(frame)\n",
      "  File \"C:\\Users\\29459\\anaconda3\\envs\\python37\\lib\\inspect.py\", line 696, in getsourcefile\n",
      "    if getattr(getmodule(object, filename), '__loader__', None) is not None:\n",
      "  File \"C:\\Users\\29459\\anaconda3\\envs\\python37\\lib\\inspect.py\", line 733, in getmodule\n",
      "    if ismodule(module) and hasattr(module, '__file__'):\n",
      "  File \"C:\\Users\\29459\\anaconda3\\envs\\python37\\lib\\site-packages\\tensorflow\\__init__.py\", line 50, in __getattr__\n",
      "    module = self._load()\n",
      "  File \"C:\\Users\\29459\\anaconda3\\envs\\python37\\lib\\site-packages\\tensorflow\\__init__.py\", line 44, in _load\n",
      "    module = _importlib.import_module(self.__name__)\n",
      "  File \"C:\\Users\\29459\\anaconda3\\envs\\python37\\lib\\importlib\\__init__.py\", line 127, in import_module\n",
      "    return _bootstrap._gcd_import(name[level:], package, level)\n",
      "  File \"<frozen importlib._bootstrap>\", line 1006, in _gcd_import\n",
      "  File \"<frozen importlib._bootstrap>\", line 983, in _find_and_load\n",
      "  File \"<frozen importlib._bootstrap>\", line 965, in _find_and_load_unlocked\n",
      "ModuleNotFoundError: No module named 'tensorflow_core.estimator'\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\29459\\anaconda3\\envs\\python37\\lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 3437, in run_code\n",
      "    exec(code_obj, self.user_global_ns, self.user_ns)\n",
      "  File \"<ipython-input-22-08ba37669687>\", line 1, in <module>\n",
      "    model.save('mnist_cnn.h5')\n",
      "  File \"C:\\Users\\29459\\anaconda3\\envs\\python37\\lib\\site-packages\\keras\\engine\\network.py\", line 1152, in save\n",
      "    save_model(self, filepath, overwrite, include_optimizer)\n",
      "  File \"C:\\Users\\29459\\anaconda3\\envs\\python37\\lib\\site-packages\\keras\\engine\\saving.py\", line 449, in save_wrapper\n",
      "    save_function(obj, filepath, overwrite, *args, **kwargs)\n",
      "  File \"C:\\Users\\29459\\anaconda3\\envs\\python37\\lib\\site-packages\\keras\\engine\\saving.py\", line 540, in save_model\n",
      "    with H5Dict(filepath, mode='w') as h5dict:\n",
      "  File \"C:\\Users\\29459\\anaconda3\\envs\\python37\\lib\\site-packages\\keras\\utils\\io_utils.py\", line 191, in __init__\n",
      "    self.data = h5py.File(path, mode=mode)\n",
      "  File \"C:\\Users\\29459\\AppData\\Roaming\\Python\\Python37\\site-packages\\h5py\\_hl\\files.py\", line 408, in __init__\n",
      "    swmr=swmr)\n",
      "  File \"C:\\Users\\29459\\AppData\\Roaming\\Python\\Python37\\site-packages\\h5py\\_hl\\files.py\", line 179, in make_fid\n",
      "    fid = h5f.create(name, h5f.ACC_TRUNC, fapl=fapl, fcpl=fcpl)\n",
      "  File \"h5py\\_objects.pyx\", line 54, in h5py._objects.with_phil.wrapper\n",
      "  File \"h5py\\_objects.pyx\", line 55, in h5py._objects.with_phil.wrapper\n",
      "  File \"h5py\\h5f.pyx\", line 108, in h5py.h5f.create\n",
      "OSError: Unable to create file (unable to truncate a file which is already open)\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\29459\\anaconda3\\envs\\python37\\lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 2061, in showtraceback\n",
      "    stb = value._render_traceback_()\n",
      "AttributeError: 'OSError' object has no attribute '_render_traceback_'\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\29459\\anaconda3\\envs\\python37\\lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 3357, in run_ast_nodes\n",
      "    if (await self.run_code(code, result,  async_=asy)):\n",
      "  File \"C:\\Users\\29459\\anaconda3\\envs\\python37\\lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 3454, in run_code\n",
      "    self.showtraceback(running_compiled_code=True)\n",
      "  File \"C:\\Users\\29459\\anaconda3\\envs\\python37\\lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 2064, in showtraceback\n",
      "    value, tb, tb_offset=tb_offset)\n",
      "  File \"C:\\Users\\29459\\anaconda3\\envs\\python37\\lib\\site-packages\\IPython\\core\\ultratb.py\", line 1368, in structured_traceback\n",
      "    self, etype, value, tb, tb_offset, number_of_lines_of_context)\n",
      "  File \"C:\\Users\\29459\\anaconda3\\envs\\python37\\lib\\site-packages\\IPython\\core\\ultratb.py\", line 1268, in structured_traceback\n",
      "    self, etype, value, tb, tb_offset, number_of_lines_of_context\n",
      "  File \"C:\\Users\\29459\\anaconda3\\envs\\python37\\lib\\site-packages\\IPython\\core\\ultratb.py\", line 1125, in structured_traceback\n",
      "    tb_offset)\n",
      "  File \"C:\\Users\\29459\\anaconda3\\envs\\python37\\lib\\site-packages\\IPython\\core\\ultratb.py\", line 1082, in format_exception_as_a_whole\n",
      "    last_unique, recursion_repeat = find_recursion(orig_etype, evalue, records)\n",
      "  File \"C:\\Users\\29459\\anaconda3\\envs\\python37\\lib\\site-packages\\IPython\\core\\ultratb.py\", line 382, in find_recursion\n",
      "    return len(records), 0\n",
      "TypeError: object of type 'NoneType' has no len()\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\29459\\anaconda3\\envs\\python37\\lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 2061, in showtraceback\n",
      "    stb = value._render_traceback_()\n",
      "AttributeError: 'TypeError' object has no attribute '_render_traceback_'\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\29459\\anaconda3\\envs\\python37\\lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 2940, in _run_cell\n",
      "    return runner(coro)\n",
      "  File \"C:\\Users\\29459\\anaconda3\\envs\\python37\\lib\\site-packages\\IPython\\core\\async_helpers.py\", line 68, in _pseudo_sync_runner\n",
      "    coro.send(None)\n",
      "  File \"C:\\Users\\29459\\anaconda3\\envs\\python37\\lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 3166, in run_cell_async\n",
      "    interactivity=interactivity, compiler=compiler, result=result)\n",
      "  File \"C:\\Users\\29459\\anaconda3\\envs\\python37\\lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 3376, in run_ast_nodes\n",
      "    self.showtraceback()\n",
      "  File \"C:\\Users\\29459\\anaconda3\\envs\\python37\\lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 2064, in showtraceback\n",
      "    value, tb, tb_offset=tb_offset)\n",
      "  File \"C:\\Users\\29459\\anaconda3\\envs\\python37\\lib\\site-packages\\IPython\\core\\ultratb.py\", line 1368, in structured_traceback\n",
      "    self, etype, value, tb, tb_offset, number_of_lines_of_context)\n",
      "  File \"C:\\Users\\29459\\anaconda3\\envs\\python37\\lib\\site-packages\\IPython\\core\\ultratb.py\", line 1268, in structured_traceback\n",
      "    self, etype, value, tb, tb_offset, number_of_lines_of_context\n",
      "  File \"C:\\Users\\29459\\anaconda3\\envs\\python37\\lib\\site-packages\\IPython\\core\\ultratb.py\", line 1143, in structured_traceback\n",
      "    chained_exceptions_tb_offset)\n",
      "  File \"C:\\Users\\29459\\anaconda3\\envs\\python37\\lib\\site-packages\\IPython\\core\\ultratb.py\", line 1082, in format_exception_as_a_whole\n",
      "    last_unique, recursion_repeat = find_recursion(orig_etype, evalue, records)\n",
      "  File \"C:\\Users\\29459\\anaconda3\\envs\\python37\\lib\\site-packages\\IPython\\core\\ultratb.py\", line 382, in find_recursion\n",
      "    return len(records), 0\n",
      "TypeError: object of type 'NoneType' has no len()\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\29459\\anaconda3\\envs\\python37\\lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 2061, in showtraceback\n",
      "    stb = value._render_traceback_()\n",
      "AttributeError: 'TypeError' object has no attribute '_render_traceback_'\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\29459\\anaconda3\\envs\\python37\\lib\\site-packages\\IPython\\core\\ultratb.py\", line 1101, in get_records\n",
      "    return _fixed_getinnerframes(etb, number_of_lines_of_context, tb_offset)\n",
      "  File \"C:\\Users\\29459\\anaconda3\\envs\\python37\\lib\\site-packages\\IPython\\core\\ultratb.py\", line 248, in wrapped\n",
      "    return f(*args, **kwargs)\n",
      "  File \"C:\\Users\\29459\\anaconda3\\envs\\python37\\lib\\site-packages\\IPython\\core\\ultratb.py\", line 281, in _fixed_getinnerframes\n",
      "    records = fix_frame_records_filenames(inspect.getinnerframes(etb, context))\n",
      "  File \"C:\\Users\\29459\\anaconda3\\envs\\python37\\lib\\inspect.py\", line 1502, in getinnerframes\n",
      "    frameinfo = (tb.tb_frame,) + getframeinfo(tb, context)\n",
      "  File \"C:\\Users\\29459\\anaconda3\\envs\\python37\\lib\\inspect.py\", line 1460, in getframeinfo\n",
      "    filename = getsourcefile(frame) or getfile(frame)\n",
      "  File \"C:\\Users\\29459\\anaconda3\\envs\\python37\\lib\\inspect.py\", line 696, in getsourcefile\n",
      "    if getattr(getmodule(object, filename), '__loader__', None) is not None:\n",
      "  File \"C:\\Users\\29459\\anaconda3\\envs\\python37\\lib\\inspect.py\", line 733, in getmodule\n",
      "    if ismodule(module) and hasattr(module, '__file__'):\n",
      "  File \"C:\\Users\\29459\\anaconda3\\envs\\python37\\lib\\site-packages\\tensorflow\\__init__.py\", line 50, in __getattr__\n",
      "    module = self._load()\n",
      "  File \"C:\\Users\\29459\\anaconda3\\envs\\python37\\lib\\site-packages\\tensorflow\\__init__.py\", line 44, in _load\n",
      "    module = _importlib.import_module(self.__name__)\n",
      "  File \"C:\\Users\\29459\\anaconda3\\envs\\python37\\lib\\importlib\\__init__.py\", line 127, in import_module\n",
      "    return _bootstrap._gcd_import(name[level:], package, level)\n",
      "  File \"<frozen importlib._bootstrap>\", line 1006, in _gcd_import\n",
      "  File \"<frozen importlib._bootstrap>\", line 983, in _find_and_load\n",
      "  File \"<frozen importlib._bootstrap>\", line 965, in _find_and_load_unlocked\n",
      "ModuleNotFoundError: No module named 'tensorflow_core.estimator'\n"
     ]
    }
   ],
   "source": [
    "model.save('mnist_cnn.h5')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "4cdfb93e",
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "\n",
    "from keras.models import load_model\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "75c81c3f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ERROR! Session/line number was not unique in database. History logging moved to new session 296\n"
     ]
    }
   ],
   "source": [
    "model.load_weights('mnist_cnn.h5')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "7470c755",
   "metadata": {},
   "outputs": [],
   "source": [
    "import h5py\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "05de689f",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\29459\\anaconda3\\envs\\python37\\lib\\site-packages\\ipykernel_launcher.py:1: H5pyDeprecationWarning: The default file mode will change to 'r' (read-only) in h5py 3.0. To suppress this warning, pass the mode you need to h5py.File(), or set the global default h5.get_config().default_file_mode, or set the environment variable H5PY_DEFAULT_READONLY=1. Available modes are: 'r', 'r+', 'w', 'w-'/'x', 'a'. See the docs for details.\n",
      "  \"\"\"Entry point for launching an IPython kernel.\n"
     ]
    }
   ],
   "source": [
    "file=h5py.File('mnist_cnn.h5')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "7007ef2b",
   "metadata": {},
   "outputs": [],
   "source": [
    "model=keras.models.load_model(file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "cc541026",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "10000/10000 [==============================] - 6s 591us/step\n",
      "[0.04901809926041169, 0.9842000007629395]\n"
     ]
    }
   ],
   "source": [
    "#评估模型\n",
    "score = model.evaluate(X_test, y_test)\n",
    "print(score)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "85d6a738",
   "metadata": {},
   "outputs": [],
   "source": [
    "from tensorflow.keras.preprocessing import image\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "4a107096",
   "metadata": {},
   "outputs": [],
   "source": [
    "img = image.load_img('./data/1.png', target_size=(224, 224))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "91753a2a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAOAAAADgCAIAAACVT/22AAAC1UlEQVR4nO3dsWrDMBRA0ar0/3/ZXTLWBgeZXKnnjCEED5c3PBxpHMfxBVXfn34AuCJQ0gRKmkBJEyhpAiVNoKQJlDSBkiZQ0gRKmkBJEyhpAiVNoKQJlDSBkiZQ0gRKmkBJEyhpAiVNoKT9fPoBKsYYf35+dm7A2ffvci7BNROUNIGSJlDSBEqaQEkTKGkCJe3f7UHv7i9n7Tt5jwlKmkBJEyhpAiVNoKQJlDSBkrbtHtT+cg8mKGkCJU2gpAmUNIGSJlDSBEra8ntQ+869maCkCZQ0gZImUNIESppASRMoacvvQZ/29PmgXDNBSRMoaQIlTaCkCZQ0gZImUNLsQV/sO5tMUNIESppASRMoaQIlTaCkCZS05feg7lvfmwlKmkBJEyhpAiVNoKQJlDSBkiZQ0gRKmkBJEyhpAiVNoKQJlDSBkrb8+6Cr8N7qe0xQ0gRKmkBJEyhpAiVNoKQJlDSBkiZQ0gRKmkBJEyhpAiVNoKQJlDSBkiZQ0gRKmkBJEyhpAiVNoKQJlDT/i39xL3yTCUqaQEkTKGkCJU2gpAmUNIGSZg86mXNA5zJBSRMoaQIlTaCkCZQ0gZImUNIESppASRMoaQIlTaCkCZQ0gZImUNIESppASRMoaQIlTaCkCZQ0gZImUNIESppASRMoaQIlTaCkCZQ0gZImUNIESppASRMoaQIlTaCkCZQ0gZImUNIESppASRMoaQIlTaCkCZQ0gZImUNLGrPvNxxhTfoe5Vr+/3gQlTaCkCZQ0gZImUNIESppASbu9B7Xv3MMq+1ETlDSBkiZQ0gRKmkBJEyhpAiXN+6CbW2XfecYEJU2gpAmUNIGSJlDSBEqaQEmbtgeFJ5igpAmUNIGSJlDSBEqaQEkTKGkCJU2gpAmUNIGSJlDSBEqaQEkTKGkCJU2gpAmUNIGSJlDSBEqaQEkTKGkCJU2gpAmUNIGSJlDSBEqaQEkTKGm/rBAux+jR84cAAAAASUVORK5CYII=\n",
      "text/plain": [
       "<PIL.Image.Image image mode=RGB size=224x224 at 0x17630C14D08>"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "img"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b6968098",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "ERROR:root:Internal Python error in the inspect module.\n",
      "Below is the traceback from this internal error.\n",
      "\n",
      "ERROR:root:Internal Python error in the inspect module.\n",
      "Below is the traceback from this internal error.\n",
      "\n",
      "ERROR:root:Internal Python error in the inspect module.\n",
      "Below is the traceback from this internal error.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\29459\\anaconda3\\envs\\python37\\lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 3437, in run_code\n",
      "    exec(code_obj, self.user_global_ns, self.user_ns)\n",
      "  File \"<ipython-input-31-c8ec22b3e787>\", line 1, in <module>\n",
      "    import cv2\n",
      "ModuleNotFoundError: No module named 'cv2'\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\29459\\anaconda3\\envs\\python37\\lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 2061, in showtraceback\n",
      "    stb = value._render_traceback_()\n",
      "AttributeError: 'ModuleNotFoundError' object has no attribute '_render_traceback_'\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\29459\\anaconda3\\envs\\python37\\lib\\site-packages\\IPython\\core\\ultratb.py\", line 1101, in get_records\n",
      "    return _fixed_getinnerframes(etb, number_of_lines_of_context, tb_offset)\n",
      "  File \"C:\\Users\\29459\\anaconda3\\envs\\python37\\lib\\site-packages\\IPython\\core\\ultratb.py\", line 248, in wrapped\n",
      "    return f(*args, **kwargs)\n",
      "  File \"C:\\Users\\29459\\anaconda3\\envs\\python37\\lib\\site-packages\\IPython\\core\\ultratb.py\", line 281, in _fixed_getinnerframes\n",
      "    records = fix_frame_records_filenames(inspect.getinnerframes(etb, context))\n",
      "  File \"C:\\Users\\29459\\anaconda3\\envs\\python37\\lib\\inspect.py\", line 1502, in getinnerframes\n",
      "    frameinfo = (tb.tb_frame,) + getframeinfo(tb, context)\n",
      "  File \"C:\\Users\\29459\\anaconda3\\envs\\python37\\lib\\inspect.py\", line 1460, in getframeinfo\n",
      "    filename = getsourcefile(frame) or getfile(frame)\n",
      "  File \"C:\\Users\\29459\\anaconda3\\envs\\python37\\lib\\inspect.py\", line 696, in getsourcefile\n",
      "    if getattr(getmodule(object, filename), '__loader__', None) is not None:\n",
      "  File \"C:\\Users\\29459\\anaconda3\\envs\\python37\\lib\\inspect.py\", line 733, in getmodule\n",
      "    if ismodule(module) and hasattr(module, '__file__'):\n",
      "  File \"C:\\Users\\29459\\anaconda3\\envs\\python37\\lib\\site-packages\\tensorflow\\__init__.py\", line 50, in __getattr__\n",
      "    module = self._load()\n",
      "  File \"C:\\Users\\29459\\anaconda3\\envs\\python37\\lib\\site-packages\\tensorflow\\__init__.py\", line 44, in _load\n",
      "    module = _importlib.import_module(self.__name__)\n",
      "  File \"C:\\Users\\29459\\anaconda3\\envs\\python37\\lib\\importlib\\__init__.py\", line 127, in import_module\n",
      "    return _bootstrap._gcd_import(name[level:], package, level)\n",
      "  File \"<frozen importlib._bootstrap>\", line 1006, in _gcd_import\n",
      "  File \"<frozen importlib._bootstrap>\", line 983, in _find_and_load\n",
      "  File \"<frozen importlib._bootstrap>\", line 965, in _find_and_load_unlocked\n",
      "ModuleNotFoundError: No module named 'tensorflow_core.estimator'\n",
      "ERROR! Session/line number was not unique in database. History logging moved to new session 297\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\29459\\anaconda3\\envs\\python37\\lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 3437, in run_code\n",
      "    exec(code_obj, self.user_global_ns, self.user_ns)\n",
      "  File \"<ipython-input-31-c8ec22b3e787>\", line 1, in <module>\n",
      "    import cv2\n",
      "ModuleNotFoundError: No module named 'cv2'\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\29459\\anaconda3\\envs\\python37\\lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 2061, in showtraceback\n",
      "    stb = value._render_traceback_()\n",
      "AttributeError: 'ModuleNotFoundError' object has no attribute '_render_traceback_'\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\29459\\anaconda3\\envs\\python37\\lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 3357, in run_ast_nodes\n",
      "    if (await self.run_code(code, result,  async_=asy)):\n",
      "  File \"C:\\Users\\29459\\anaconda3\\envs\\python37\\lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 3454, in run_code\n",
      "    self.showtraceback(running_compiled_code=True)\n",
      "  File \"C:\\Users\\29459\\anaconda3\\envs\\python37\\lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 2064, in showtraceback\n",
      "    value, tb, tb_offset=tb_offset)\n",
      "  File \"C:\\Users\\29459\\anaconda3\\envs\\python37\\lib\\site-packages\\IPython\\core\\ultratb.py\", line 1368, in structured_traceback\n",
      "    self, etype, value, tb, tb_offset, number_of_lines_of_context)\n",
      "  File \"C:\\Users\\29459\\anaconda3\\envs\\python37\\lib\\site-packages\\IPython\\core\\ultratb.py\", line 1268, in structured_traceback\n",
      "    self, etype, value, tb, tb_offset, number_of_lines_of_context\n",
      "  File \"C:\\Users\\29459\\anaconda3\\envs\\python37\\lib\\site-packages\\IPython\\core\\ultratb.py\", line 1125, in structured_traceback\n",
      "    tb_offset)\n",
      "  File \"C:\\Users\\29459\\anaconda3\\envs\\python37\\lib\\site-packages\\IPython\\core\\ultratb.py\", line 1082, in format_exception_as_a_whole\n",
      "    last_unique, recursion_repeat = find_recursion(orig_etype, evalue, records)\n",
      "  File \"C:\\Users\\29459\\anaconda3\\envs\\python37\\lib\\site-packages\\IPython\\core\\ultratb.py\", line 382, in find_recursion\n",
      "    return len(records), 0\n",
      "TypeError: object of type 'NoneType' has no len()\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\29459\\anaconda3\\envs\\python37\\lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 2061, in showtraceback\n",
      "    stb = value._render_traceback_()\n",
      "AttributeError: 'TypeError' object has no attribute '_render_traceback_'\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\29459\\anaconda3\\envs\\python37\\lib\\site-packages\\IPython\\core\\ultratb.py\", line 1101, in get_records\n",
      "    return _fixed_getinnerframes(etb, number_of_lines_of_context, tb_offset)\n",
      "  File \"C:\\Users\\29459\\anaconda3\\envs\\python37\\lib\\site-packages\\IPython\\core\\ultratb.py\", line 248, in wrapped\n",
      "    return f(*args, **kwargs)\n",
      "  File \"C:\\Users\\29459\\anaconda3\\envs\\python37\\lib\\site-packages\\IPython\\core\\ultratb.py\", line 281, in _fixed_getinnerframes\n",
      "    records = fix_frame_records_filenames(inspect.getinnerframes(etb, context))\n",
      "  File \"C:\\Users\\29459\\anaconda3\\envs\\python37\\lib\\inspect.py\", line 1502, in getinnerframes\n",
      "    frameinfo = (tb.tb_frame,) + getframeinfo(tb, context)\n",
      "  File \"C:\\Users\\29459\\anaconda3\\envs\\python37\\lib\\inspect.py\", line 1460, in getframeinfo\n",
      "    filename = getsourcefile(frame) or getfile(frame)\n",
      "  File \"C:\\Users\\29459\\anaconda3\\envs\\python37\\lib\\inspect.py\", line 696, in getsourcefile\n",
      "    if getattr(getmodule(object, filename), '__loader__', None) is not None:\n",
      "  File \"C:\\Users\\29459\\anaconda3\\envs\\python37\\lib\\inspect.py\", line 733, in getmodule\n",
      "    if ismodule(module) and hasattr(module, '__file__'):\n",
      "  File \"C:\\Users\\29459\\anaconda3\\envs\\python37\\lib\\site-packages\\tensorflow\\__init__.py\", line 50, in __getattr__\n",
      "    module = self._load()\n",
      "  File \"C:\\Users\\29459\\anaconda3\\envs\\python37\\lib\\site-packages\\tensorflow\\__init__.py\", line 44, in _load\n",
      "    module = _importlib.import_module(self.__name__)\n",
      "  File \"C:\\Users\\29459\\anaconda3\\envs\\python37\\lib\\importlib\\__init__.py\", line 127, in import_module\n",
      "    return _bootstrap._gcd_import(name[level:], package, level)\n",
      "  File \"<frozen importlib._bootstrap>\", line 1006, in _gcd_import\n",
      "  File \"<frozen importlib._bootstrap>\", line 983, in _find_and_load\n",
      "  File \"<frozen importlib._bootstrap>\", line 965, in _find_and_load_unlocked\n",
      "ModuleNotFoundError: No module named 'tensorflow_core.estimator'\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\29459\\anaconda3\\envs\\python37\\lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 3437, in run_code\n",
      "    exec(code_obj, self.user_global_ns, self.user_ns)\n",
      "  File \"<ipython-input-31-c8ec22b3e787>\", line 1, in <module>\n",
      "    import cv2\n",
      "ModuleNotFoundError: No module named 'cv2'\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\29459\\anaconda3\\envs\\python37\\lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 2061, in showtraceback\n",
      "    stb = value._render_traceback_()\n",
      "AttributeError: 'ModuleNotFoundError' object has no attribute '_render_traceback_'\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\29459\\anaconda3\\envs\\python37\\lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 3357, in run_ast_nodes\n",
      "    if (await self.run_code(code, result,  async_=asy)):\n",
      "  File \"C:\\Users\\29459\\anaconda3\\envs\\python37\\lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 3454, in run_code\n",
      "    self.showtraceback(running_compiled_code=True)\n",
      "  File \"C:\\Users\\29459\\anaconda3\\envs\\python37\\lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 2064, in showtraceback\n",
      "    value, tb, tb_offset=tb_offset)\n",
      "  File \"C:\\Users\\29459\\anaconda3\\envs\\python37\\lib\\site-packages\\IPython\\core\\ultratb.py\", line 1368, in structured_traceback\n",
      "    self, etype, value, tb, tb_offset, number_of_lines_of_context)\n",
      "  File \"C:\\Users\\29459\\anaconda3\\envs\\python37\\lib\\site-packages\\IPython\\core\\ultratb.py\", line 1268, in structured_traceback\n",
      "    self, etype, value, tb, tb_offset, number_of_lines_of_context\n",
      "  File \"C:\\Users\\29459\\anaconda3\\envs\\python37\\lib\\site-packages\\IPython\\core\\ultratb.py\", line 1125, in structured_traceback\n",
      "    tb_offset)\n",
      "  File \"C:\\Users\\29459\\anaconda3\\envs\\python37\\lib\\site-packages\\IPython\\core\\ultratb.py\", line 1082, in format_exception_as_a_whole\n",
      "    last_unique, recursion_repeat = find_recursion(orig_etype, evalue, records)\n",
      "  File \"C:\\Users\\29459\\anaconda3\\envs\\python37\\lib\\site-packages\\IPython\\core\\ultratb.py\", line 382, in find_recursion\n",
      "    return len(records), 0\n",
      "TypeError: object of type 'NoneType' has no len()\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\29459\\anaconda3\\envs\\python37\\lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 2061, in showtraceback\n",
      "    stb = value._render_traceback_()\n",
      "AttributeError: 'TypeError' object has no attribute '_render_traceback_'\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\29459\\anaconda3\\envs\\python37\\lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 2940, in _run_cell\n",
      "    return runner(coro)\n",
      "  File \"C:\\Users\\29459\\anaconda3\\envs\\python37\\lib\\site-packages\\IPython\\core\\async_helpers.py\", line 68, in _pseudo_sync_runner\n",
      "    coro.send(None)\n",
      "  File \"C:\\Users\\29459\\anaconda3\\envs\\python37\\lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 3166, in run_cell_async\n",
      "    interactivity=interactivity, compiler=compiler, result=result)\n",
      "  File \"C:\\Users\\29459\\anaconda3\\envs\\python37\\lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 3376, in run_ast_nodes\n",
      "    self.showtraceback()\n",
      "  File \"C:\\Users\\29459\\anaconda3\\envs\\python37\\lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 2064, in showtraceback\n",
      "    value, tb, tb_offset=tb_offset)\n",
      "  File \"C:\\Users\\29459\\anaconda3\\envs\\python37\\lib\\site-packages\\IPython\\core\\ultratb.py\", line 1368, in structured_traceback\n",
      "    self, etype, value, tb, tb_offset, number_of_lines_of_context)\n",
      "  File \"C:\\Users\\29459\\anaconda3\\envs\\python37\\lib\\site-packages\\IPython\\core\\ultratb.py\", line 1268, in structured_traceback\n",
      "    self, etype, value, tb, tb_offset, number_of_lines_of_context\n",
      "  File \"C:\\Users\\29459\\anaconda3\\envs\\python37\\lib\\site-packages\\IPython\\core\\ultratb.py\", line 1143, in structured_traceback\n",
      "    chained_exceptions_tb_offset)\n",
      "  File \"C:\\Users\\29459\\anaconda3\\envs\\python37\\lib\\site-packages\\IPython\\core\\ultratb.py\", line 1082, in format_exception_as_a_whole\n",
      "    last_unique, recursion_repeat = find_recursion(orig_etype, evalue, records)\n",
      "  File \"C:\\Users\\29459\\anaconda3\\envs\\python37\\lib\\site-packages\\IPython\\core\\ultratb.py\", line 382, in find_recursion\n",
      "    return len(records), 0\n",
      "TypeError: object of type 'NoneType' has no len()\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\29459\\anaconda3\\envs\\python37\\lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 2061, in showtraceback\n",
      "    stb = value._render_traceback_()\n",
      "AttributeError: 'TypeError' object has no attribute '_render_traceback_'\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\29459\\anaconda3\\envs\\python37\\lib\\site-packages\\IPython\\core\\ultratb.py\", line 1101, in get_records\n",
      "    return _fixed_getinnerframes(etb, number_of_lines_of_context, tb_offset)\n",
      "  File \"C:\\Users\\29459\\anaconda3\\envs\\python37\\lib\\site-packages\\IPython\\core\\ultratb.py\", line 248, in wrapped\n",
      "    return f(*args, **kwargs)\n",
      "  File \"C:\\Users\\29459\\anaconda3\\envs\\python37\\lib\\site-packages\\IPython\\core\\ultratb.py\", line 281, in _fixed_getinnerframes\n",
      "    records = fix_frame_records_filenames(inspect.getinnerframes(etb, context))\n",
      "  File \"C:\\Users\\29459\\anaconda3\\envs\\python37\\lib\\inspect.py\", line 1502, in getinnerframes\n",
      "    frameinfo = (tb.tb_frame,) + getframeinfo(tb, context)\n",
      "  File \"C:\\Users\\29459\\anaconda3\\envs\\python37\\lib\\inspect.py\", line 1460, in getframeinfo\n",
      "    filename = getsourcefile(frame) or getfile(frame)\n",
      "  File \"C:\\Users\\29459\\anaconda3\\envs\\python37\\lib\\inspect.py\", line 696, in getsourcefile\n",
      "    if getattr(getmodule(object, filename), '__loader__', None) is not None:\n",
      "  File \"C:\\Users\\29459\\anaconda3\\envs\\python37\\lib\\inspect.py\", line 733, in getmodule\n",
      "    if ismodule(module) and hasattr(module, '__file__'):\n",
      "  File \"C:\\Users\\29459\\anaconda3\\envs\\python37\\lib\\site-packages\\tensorflow\\__init__.py\", line 50, in __getattr__\n",
      "    module = self._load()\n",
      "  File \"C:\\Users\\29459\\anaconda3\\envs\\python37\\lib\\site-packages\\tensorflow\\__init__.py\", line 44, in _load\n",
      "    module = _importlib.import_module(self.__name__)\n",
      "  File \"C:\\Users\\29459\\anaconda3\\envs\\python37\\lib\\importlib\\__init__.py\", line 127, in import_module\n",
      "    return _bootstrap._gcd_import(name[level:], package, level)\n",
      "  File \"<frozen importlib._bootstrap>\", line 1006, in _gcd_import\n",
      "  File \"<frozen importlib._bootstrap>\", line 983, in _find_and_load\n",
      "  File \"<frozen importlib._bootstrap>\", line 965, in _find_and_load_unlocked\n",
      "ModuleNotFoundError: No module named 'tensorflow_core.estimator'\n"
     ]
    }
   ],
   "source": [
    "def _decode_and_resize(filename, label):\n",
    "    image_string = tf.io.read_file(filename) \n",
    "    image_decoded = tf.image.decode_jpeg(image_string) \n",
    "    image_resized = tf.image.resize(image_decoded, [256, 256]) / 255.0 \n",
    "    return image_resized, label"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c9ddc69b",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
